{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e2af07ff",
   "metadata": {},
   "source": [
    "# Getting Started with Arrays in BigQuery\n",
    "Arrays are an important topic to understand on at least a basic level to work with data in BigQuery. It allows additional information to be stored in lists or key-value pairs within the table. Both can be searched on and extracted to find exactly what you are looking for.\n",
    "\n",
    "First you need to install and prepare the environment again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ee747a28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting google.cloud.bigquery\n",
      "  Using cached https://files.pythonhosted.org/packages/31/b5/f9deb0ceb925fe04afdfb4d2258ef64959e22e45468432f5b97caf1b85ba/google_cloud_bigquery-2.26.0-py2.py3-none-any.whl\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.19.0)\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.7.2)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (3.17.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (2.22.0)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.39.0)\n",
      "Requirement already satisfied: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.31.1)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.3.3)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (21.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (1.16.0)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (1.34.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2021.5.30)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (1.25.11)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2.8)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (41.0.1)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (2021.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (1.53.0)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.1.2)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google.cloud.bigquery) (2.4.7)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.2.2)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.7.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.2.8)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.12.3)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.4.8)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (2.19)\n",
      "Installing collected packages: google.cloud.bigquery\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_bigquery-2.26.0-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google.cloud.storage\n",
      "  Using cached https://files.pythonhosted.org/packages/76/03/55d5255ce6226d8908d5354f3d2ab5749a0ec8b75efc9cdd7d8a45db17b0/google_cloud_storage-1.42.1-py2.py3-none-any.whl\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.3.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.storage) (2.22.0)\n",
      "Requirement already satisfied: google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.34.0)\n",
      "Requirement already satisfied: google-cloud-core<3.0dev,>=1.6.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.7.2)\n",
      "Requirement already satisfied: google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.31.1)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.1.2)\n",
      "Requirement already satisfied: six>=1.4.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.16.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2021.5.30)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2.8)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (1.25.11)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (3.0.4)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.7.2)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.2.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.2.8)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (41.0.1)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (3.17.3)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (1.53.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (21.0)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2021.1)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.12.3)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.4.8)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2.4.7)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (2.19)\n",
      "Installing collected packages: google.cloud.storage\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_storage-1.42.1-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google-cloud-bigquery\n",
      "\u001b[33m  Cache entry deserialization failed, entry ignored\u001b[0m\n",
      "  Using cached https://files.pythonhosted.org/packages/31/b5/f9deb0ceb925fe04afdfb4d2258ef64959e22e45468432f5b97caf1b85ba/google_cloud_bigquery-2.26.0-py2.py3-none-any.whl\n",
      "Collecting google-resumable-media<3.0dev,>=0.6.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/91/77/9e3a4b7098982a5d3ba7bd81e98436c0dc3d368f6d4576f3733ecbf5dbe1/google_resumable_media-2.0.2-py2.py3-none-any.whl\n",
      "Collecting protobuf>=3.12.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/53/4e/e2db88d0bb0bda6a879eea62fddbaf813719ce3770d458bc5580512d9c95/protobuf-3.17.3-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.whl\n",
      "Collecting google-api-core[grpc]<3.0.0dev,>=1.29.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8f/3c/6c64ce599505b5007e3b91ce9947030397f2bff591f52dbf708a0ccca4e1/google_api_core-2.0.1-py2.py3-none-any.whl\n",
      "Collecting requests<3.0.0dev,>=2.18.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/92/96/144f70b972a9c0eabbd4391ef93ccd49d0f2747f4f6a2a2738e99e5adc65/requests-2.26.0-py2.py3-none-any.whl\n",
      "Collecting packaging>=14.3 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/3c/77/e2362b676dc5008d81be423070dd9577fa03be5da2ba1105811900fda546/packaging-21.0-py3-none-any.whl\n",
      "Collecting google-cloud-core<3.0.0dev,>=1.4.1 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/42/b8/4ad3e423105d26a0c5eea54320aed57b7be060d1cd9faea53da713b408c6/google_cloud_core-2.0.0-py2.py3-none-any.whl\n",
      "Collecting proto-plus>=1.10.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8c/72/6f3f4cdc5bb0294f8d7f3f8aacb617b4c3cb17554ed78f7e28009162c795/proto_plus-1.19.0-py3-none-any.whl\n",
      "Collecting grpcio<2.0dev,>=1.38.1 (from google-cloud-bigquery)\n",
      "Collecting google-crc32c<=1.1.2,>=1.0 (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/5c/81/b71c645bac55d5c8f2d6e90603007c2940521ed63b177b833d1cd3ba7df8/google_crc32c-1.1.2-cp36-cp36m-manylinux1_x86_64.whl\n",
      "Collecting six>=1.9 (from protobuf>=3.12.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl\n",
      "Collecting setuptools>=40.3.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/c4/c1/aed7dfedb18ea73d7713bf6ca034ab001a6425be49ffa7e79bbd5999f677/setuptools-58.0.4-py3-none-any.whl\n",
      "Collecting google-auth<3.0dev,>=1.25.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/48/27/4e5c7edb04bd5bb19fd1ab6add5a429f54bc45676ba93ccfd4e5331ba928/google_auth-2.0.2-py2.py3-none-any.whl\n",
      "Collecting googleapis-common-protos<2.0dev,>=1.6.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/55/08/796a6bc0b550e2b7116041c953d3d5100016abea106131d71e5651826e7b/googleapis_common_protos-1.53.0-py2.py3-none-any.whl\n",
      "Collecting idna<4,>=2.5; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d7/77/ff688d1504cdc4db2a938e2b7b9adee5dd52e34efbd2431051efc9984de9/idna-3.2-py3-none-any.whl\n",
      "Collecting certifi>=2017.4.17 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/05/1b/0a0dece0e8aa492a6ec9e4ad2fe366b511558cdc73fd3abc82ba7348e875/certifi-2021.5.30-py2.py3-none-any.whl\n",
      "Collecting urllib3<1.27,>=1.21.1 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/5f/64/43575537846896abac0b15c3e5ac678d787a4021e906703f1766bfb8ea11/urllib3-1.26.6-py2.py3-none-any.whl\n",
      "Collecting charset-normalizer~=2.0.0; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/33/53/b7f6126a2b9fd878b025fe3c40266cfaad696f312165008ce045bffa3fe7/charset_normalizer-2.0.4-py3-none-any.whl\n",
      "Collecting pyparsing>=2.0.2 (from packaging>=14.3->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8a/bb/488841f56197b13700afd5658fc279a2025a39e22449b7cf29864669b15d/pyparsing-2.4.7-py2.py3-none-any.whl\n",
      "Collecting cffi>=1.0.0 (from google-crc32c<=1.1.2,>=1.0->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/f2/cd/3f5f059fed635d71047fa9ce507635088f982ab280fc24cde91d9afb9c1c/cffi-1.14.6-cp36-cp36m-manylinux1_x86_64.whl\n",
      "Collecting rsa<5,>=3.1.4 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/e9/93/0c0f002031f18b53af7a6166103c02b9c0667be528944137cc954ec921b3/rsa-4.7.2-py3-none-any.whl\n",
      "Collecting pyasn1-modules>=0.2.1 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/95/de/214830a981892a3e286c3794f41ae67a4495df1108c3da8a9f62159b9a9d/pyasn1_modules-0.2.8-py2.py3-none-any.whl\n",
      "Collecting cachetools<5.0,>=2.0.0 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bf/28/c4f5796c67ad06bb91d98d543a5e01805c1ff065e08871f78e52d2a331ad/cachetools-4.2.2-py3-none-any.whl\n",
      "Collecting pycparser (from cffi>=1.0.0->google-crc32c<=1.1.2,>=1.0->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/ae/e7/d9c3a176ca4b02024debf82342dab36efadfc5776f9c8db077e8f6e71821/pycparser-2.20-py2.py3-none-any.whl\n",
      "Collecting pyasn1>=0.1.3 (from rsa<5,>=3.1.4->google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/62/1e/a94a8d635fa3ce4cfc7f506003548d0a2447ae76fd5ca53932970fe3053f/pyasn1-0.4.8-py2.py3-none-any.whl\n",
      "Installing collected packages: pycparser, cffi, google-crc32c, google-resumable-media, six, protobuf, setuptools, pyasn1, rsa, pyasn1-modules, cachetools, google-auth, idna, certifi, urllib3, charset-normalizer, requests, googleapis-common-protos, grpcio, google-api-core, pyparsing, packaging, google-cloud-core, proto-plus, google-cloud-bigquery\n",
      "Successfully installed cachetools-4.2.2 certifi-2021.5.30 cffi-1.14.6 charset-normalizer-2.0.4 google-api-core-2.0.1 google-auth-2.0.2 google-cloud-bigquery-2.26.0 google-cloud-core-2.0.0 google-crc32c-1.1.2 google-resumable-media-2.0.2 googleapis-common-protos-1.53.0 grpcio-1.40.0 idna-3.2 packaging-21.0 proto-plus-1.19.0 protobuf-3.17.3 pyasn1-0.4.8 pyasn1-modules-0.2.8 pycparser-2.20 pyparsing-2.4.7 requests-2.26.0 rsa-4.7.2 setuptools-58.0.4 six-1.16.0 urllib3-1.26.6\n",
      "Requirement already satisfied: google in /home/jupyter-astine/.local/lib/python3.7/site-packages (3.0.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google) (4.9.3)\n",
      "Requirement already satisfied: soupsieve>1.2; python_version >= \"3.0\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from beautifulsoup4->google) (2.2.1)\n",
      "Requirement already up-to-date: google-cloud-bigquery[bqstorage,pandas] in /home/jupyter-astine/.local/lib/python3.7/site-packages (2.26.0)\n",
      "Requirement already satisfied, skipping upgrade: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.22.0)\n",
      "Requirement already satisfied, skipping upgrade: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (21.0)\n",
      "Requirement already satisfied, skipping upgrade: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.19.0)\n",
      "Requirement already satisfied, skipping upgrade: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.39.0)\n",
      "Requirement already satisfied, skipping upgrade: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (3.17.3)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.7.2)\n",
      "Requirement already satisfied, skipping upgrade: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.3)\n",
      "Requirement already satisfied, skipping upgrade: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.31.1)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.6.3)\n",
      "Requirement already satisfied, skipping upgrade: pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (5.0.0)\n",
      "Requirement already satisfied, skipping upgrade: pandas>=0.23.0; extra == \"pandas\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.1)\n",
      "Requirement already satisfied, skipping upgrade: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2.8)\n",
      "Requirement already satisfied, skipping upgrade: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (3.0.4)\n",
      "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (1.25.11)\n",
      "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2021.5.30)\n",
      "Requirement already satisfied, skipping upgrade: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-cloud-bigquery[bqstorage,pandas]) (2.4.7)\n",
      "Requirement already satisfied, skipping upgrade: six>=1.5.2 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from grpcio<2.0dev,>=1.38.1->google-cloud-bigquery[bqstorage,pandas]) (1.16.0)\n",
      "Requirement already satisfied, skipping upgrade: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (1.34.0)\n",
      "Requirement already satisfied, skipping upgrade: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.1.2)\n",
      "Requirement already satisfied, skipping upgrade: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (41.0.1)\n",
      "Requirement already satisfied, skipping upgrade: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (2021.1)\n",
      "Requirement already satisfied, skipping upgrade: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (1.53.0)\n",
      "Requirement already satisfied, skipping upgrade: libcst>=0.2.5 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.3.20)\n",
      "Requirement already satisfied, skipping upgrade: numpy>=1.16.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (1.21.1)\n",
      "Requirement already satisfied, skipping upgrade: python-dateutil>=2.7.3 in /opt/tljh/user/lib/python3.7/site-packages (from pandas>=0.23.0; extra == \"pandas\"->google-cloud-bigquery[bqstorage,pandas]) (2.8.2)\n",
      "Requirement already satisfied, skipping upgrade: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.2.2)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.2.8)\n",
      "Requirement already satisfied, skipping upgrade: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.7.2)\n",
      "Requirement already satisfied, skipping upgrade: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.12.3)\n",
      "Requirement already satisfied, skipping upgrade: typing-extensions>=3.7.4.2 in /opt/tljh/user/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (3.10.0.0)\n",
      "Requirement already satisfied, skipping upgrade: pyyaml>=5.2 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (5.4.1)\n",
      "Requirement already satisfied, skipping upgrade: typing-inspect>=0.4.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.7.1)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1<0.5.0,>=0.4.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.4.8)\n",
      "Requirement already satisfied, skipping upgrade: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (2.19)\n",
      "Requirement already satisfied, skipping upgrade: mypy-extensions>=0.3.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from typing-inspect>=0.4.0->libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.4.3)\n"
     ]
    }
   ],
   "source": [
    "# These commands were run when the notebook started but are shown in case you choose to work with Notebooks later.\n",
    "!pip install google.cloud.bigquery\n",
    "!pip install google.cloud.storage\n",
    "!pip3 install --upgrade google-cloud-bigquery\n",
    "!pip install google --user\n",
    "!pip install --upgrade 'google-cloud-bigquery[bqstorage,pandas]' --user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4b5fe4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this command block to install the BigQuery extension.\n",
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c5f6779",
   "metadata": {},
   "source": [
    "# Problem : I want to find RNA-Seq data from estrogen receptor positive breast cancer cell lines.\n",
    "Doing a search like this will require looking into the attributes stored in the database for each record. If you are an experienced user of SRA you might already know how to do a search like this with Entrez and Run Selector. Here we will find the results in a single SQL search primarily using the attributes column. The methods used to do this can be applied to build very targeted searches of submitter supplied information.\n",
    "\n",
    "## What Does an Array look like in BigQuery?\n",
    "Here you will look at a couple of columns from the database to show examples of arrays. You will select the accession column and the columns that are either simple arrays or are an array of data structures. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18743fb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 2/2 [00:00<00:00, 1571.49query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  6.98rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRR209136</td>\n",
       "      <td>[sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '830289504'}, {'k': 'byte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DRR209149</td>\n",
       "      <td>[sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '954170736'}, {'k': 'byte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR364422</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'geo_accession_exp', 'v': 'GSM833111'},...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR006198</td>\n",
       "      <td>[sra, srf]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bi_gssr_sample_id_exp', 'v': '24519.0'...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR003485</td>\n",
       "      <td>[sra, srf]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bi_gssr_sample_id_exp', 'v': '24516.0'...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         acc datastore_filetype datastore_provider  \\\n",
       "0  DRR209136              [sra]     [gs, ncbi, s3]   \n",
       "1  DRR209149              [sra]     [gs, ncbi, s3]   \n",
       "2  SRR364422       [fastq, sra]     [gs, ncbi, s3]   \n",
       "3  SRR006198         [sra, srf]     [gs, ncbi, s3]   \n",
       "4  SRR003485         [sra, srf]     [gs, ncbi, s3]   \n",
       "\n",
       "                     datastore_region  \\\n",
       "0  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "1  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "2  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "3  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "4  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "\n",
       "                                          attributes  \n",
       "0  [{'k': 'bases', 'v': '830289504'}, {'k': 'byte...  \n",
       "1  [{'k': 'bases', 'v': '954170736'}, {'k': 'byte...  \n",
       "2  [{'k': 'geo_accession_exp', 'v': 'GSM833111'},...  \n",
       "3  [{'k': 'bi_gssr_sample_id_exp', 'v': '24519.0'...  \n",
       "4  [{'k': 'bi_gssr_sample_id_exp', 'v': '24516.0'...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, datastore_filetype, datastore_provider, datastore_region, attributes\n",
    "FROM `nih-sra-datastore.sra.metadata`\n",
    "WHERE organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6c50bee",
   "metadata": {},
   "source": [
    "## Viewing what is in the Attributes Column of an SRA Record\n",
    "To understand what is in the attributes column in the database it may help to view the contents as a table. All of the columns other than **acc** are arrays \\[ \\] with a comma separated list of items inside. The **attributes** column is an array of structs { } that have a key (k) and a value (v) for each item in the array. You will use the UNNEST function in BigQuery to extract the contents of the attributes column into one item per row. \n",
    "\n",
    "## Listing the Contents of an Array\n",
    "If you unnest the attributes column, you get the list of the key-value pairs. The contents of this column are some standard items like \"bytes\" and \"bases\" as well as items that were provided by the submitter as additional information. This means that runs in the database will have some attributes in common but also some attributes not on other runs. There can also be keys that are repeated so one key might have multiple values in a single run. \n",
    "\n",
    "### Naming Unnested Data\n",
    "When you unnest the data you can give the result of the unnest function a name (unnested_attributes in the below query) and then use that name in the select statement. Like you did before when naming tables, it can be used to make queries a little easier to type and read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "713e06d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 584.49query/s]                          \n",
      "Downloading: 100%|██████████| 18/18 [00:00<00:00, 27.48rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>extracted_attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'k': 'sex_calc', 'v': 'female'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'k': 'bases', 'v': '12470223358'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{'k': 'bytes', 'v': '6566142480'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{'k': 'age_sam', 'v': 'NA'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{'k': 'biomaterial_provider_sam', 'v': 'Robert...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>{'k': 'cell_line_sam_ss_dpl110', 'v': 'HCA-7'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>{'k': 'isolate_sam', 'v': 'cell lines'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>{'k': 'tissue_sam', 'v': 'colon'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>{'k': 'primary_search', 'v': '304768'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>{'k': 'primary_search', 'v': '4309137'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'HCA-7'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'HCA-7-fastq'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'PRJNA304768'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SAMN04309137'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRP066923'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRR2973262'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRS1187131'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRX1459675'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 extracted_attributes\n",
       "0                    {'k': 'sex_calc', 'v': 'female'}\n",
       "1                  {'k': 'bases', 'v': '12470223358'}\n",
       "2                   {'k': 'bytes', 'v': '6566142480'}\n",
       "3                         {'k': 'age_sam', 'v': 'NA'}\n",
       "4   {'k': 'biomaterial_provider_sam', 'v': 'Robert...\n",
       "5      {'k': 'cell_line_sam_ss_dpl110', 'v': 'HCA-7'}\n",
       "6             {'k': 'isolate_sam', 'v': 'cell lines'}\n",
       "7                   {'k': 'tissue_sam', 'v': 'colon'}\n",
       "8              {'k': 'primary_search', 'v': '304768'}\n",
       "9             {'k': 'primary_search', 'v': '4309137'}\n",
       "10              {'k': 'primary_search', 'v': 'HCA-7'}\n",
       "11        {'k': 'primary_search', 'v': 'HCA-7-fastq'}\n",
       "12        {'k': 'primary_search', 'v': 'PRJNA304768'}\n",
       "13       {'k': 'primary_search', 'v': 'SAMN04309137'}\n",
       "14          {'k': 'primary_search', 'v': 'SRP066923'}\n",
       "15         {'k': 'primary_search', 'v': 'SRR2973262'}\n",
       "16         {'k': 'primary_search', 'v': 'SRS1187131'}\n",
       "17         {'k': 'primary_search', 'v': 'SRX1459675'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT unnested_attributes\n",
    "FROM `nih-sra-datastore.sra.metadata`,\n",
    "    UNNEST(attributes) as unnested_attributes \n",
    "WHERE acc = 'SRR2973262'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9962febe",
   "metadata": {},
   "source": [
    "## Listing the Keys and Values in the Attributes as a Table\n",
    "You can further break a structured list like this {'k': 'sex_calc', 'v': 'female'} apart into (k) and (v) columns by using unnested_attributes.k and unnested_attributes.v in the query. At this point you will be seeing all the keys and values in a table format for a single accession in the metadata table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d2ec3b3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 791.68query/s]                          \n",
      "Downloading: 100%|██████████| 18/18 [00:00<00:00, 26.09rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>v</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sex_calc</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bases</td>\n",
       "      <td>12470223358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bytes</td>\n",
       "      <td>6566142480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>age_sam</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "      <td>Robert Coffey, Vanderbilt University</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cell_line_sam_ss_dpl110</td>\n",
       "      <td>HCA-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>isolate_sam</td>\n",
       "      <td>cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tissue_sam</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>304768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>4309137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>HCA-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>HCA-7-fastq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>PRJNA304768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SAMN04309137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRP066923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRR2973262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRS1187131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRX1459675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           k                                     v\n",
       "0                   sex_calc                                female\n",
       "1                      bases                           12470223358\n",
       "2                      bytes                            6566142480\n",
       "3                    age_sam                                    NA\n",
       "4   biomaterial_provider_sam  Robert Coffey, Vanderbilt University\n",
       "5    cell_line_sam_ss_dpl110                                 HCA-7\n",
       "6                isolate_sam                            cell lines\n",
       "7                 tissue_sam                                 colon\n",
       "8             primary_search                                304768\n",
       "9             primary_search                               4309137\n",
       "10            primary_search                                 HCA-7\n",
       "11            primary_search                           HCA-7-fastq\n",
       "12            primary_search                           PRJNA304768\n",
       "13            primary_search                          SAMN04309137\n",
       "14            primary_search                             SRP066923\n",
       "15            primary_search                            SRR2973262\n",
       "16            primary_search                            SRS1187131\n",
       "17            primary_search                            SRX1459675"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT unnested_attributes.k, unnested_attributes.v\n",
    "FROM `nih-sra-datastore.sra.metadata`,\n",
    "    UNNEST(attributes) as unnested_attributes \n",
    "WHERE acc = 'SRR2973262'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb2de6d",
   "metadata": {},
   "source": [
    "## Finding Runs with a Certain Attribute\n",
    "Listing out all the attribute keys and values can be quite useful to help build searches for very specific things. Looking at the results for the above query, one of the attributes is tissue_sam : colon. If you want to find all runs where that exists you can use the following query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b31b4101",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 946.58query/s]                          \n",
      "Downloading: 100%|██████████| 15550/15550 [00:01<00:00, 13305.05rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR14684549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR15652167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR7762696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR7762697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR7762698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15545</th>\n",
       "      <td>SRR13345069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15546</th>\n",
       "      <td>SRR13872307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15547</th>\n",
       "      <td>SRR15347514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15548</th>\n",
       "      <td>SRR11631045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15549</th>\n",
       "      <td>SRR5047917</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15550 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               acc\n",
       "0      SRR14684549\n",
       "1      SRR15652167\n",
       "2       SRR7762696\n",
       "3       SRR7762697\n",
       "4       SRR7762698\n",
       "...            ...\n",
       "15545  SRR13345069\n",
       "15546  SRR13872307\n",
       "15547  SRR15347514\n",
       "15548  SRR11631045\n",
       "15549   SRR5047917\n",
       "\n",
       "[15550 rows x 1 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc\n",
    "FROM `nih-sra-datastore.sra.metadata`   \n",
    "WHERE ('tissue_sam', 'colon') in UNNEST(attributes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca8d90c1",
   "metadata": {},
   "source": [
    "## Search for Breast Cancer Cell Line Records\n",
    "Now you can use the UNNEST function to search for the data in the initial problem. The LIKE function with the wildcard % in will let you search for records that contain certain words or phrases with anything around them. The query below will search for all records that mention \"breast\" and \"cancer\" and \"cell\" and \"line\" in at least one attribute."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "86a15a96",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 1042.84query/s]                         \n",
      "Downloading: 100%|██████████| 1894/1894 [00:01<00:00, 1683.96rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR15443831</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA754393</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8454880</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA515879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR2353162</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA295409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR3290939</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA316332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR3290972</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA316332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1889</th>\n",
       "      <td>SRR13036294</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA676130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1890</th>\n",
       "      <td>SRR12060822</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA640820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>SRR12060836</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA640820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1892</th>\n",
       "      <td>SRR6822820</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA437670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>SRR6822846</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA437670</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1894 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              acc assay_type consent   librarysource libraryselection  \\\n",
       "0     SRR15443831    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1      SRR8454880    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "2      SRR2353162    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "3      SRR3290939    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "4      SRR3290972    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "...           ...        ...     ...             ...              ...   \n",
       "1889  SRR13036294    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1890  SRR12060822    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1891  SRR12060836    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1892   SRR6822820    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1893   SRR6822846    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "\n",
       "       bioproject  \n",
       "0     PRJNA754393  \n",
       "1     PRJNA515879  \n",
       "2     PRJNA295409  \n",
       "3     PRJNA316332  \n",
       "4     PRJNA316332  \n",
       "...           ...  \n",
       "1889  PRJNA676130  \n",
       "1890  PRJNA640820  \n",
       "1891  PRJNA640820  \n",
       "1892  PRJNA437670  \n",
       "1893  PRJNA437670  \n",
       "\n",
       "[1894 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE extracted.v LIKE '%breast%cancer%cell%line%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec96c153",
   "metadata": {},
   "source": [
    "### Subselect\n",
    "A subselect is a select statement embedded within another select statement. In this case you will use a subselect to get the value of attributes using the attribute key name. This example uses 'source_name_sam' to see values that are on *some* records and 'bases' which will be on all records.\n",
    "\n",
    "You can also find the key for the attribute that contained the breast cancer cell line entry by using a subselect. Because there might be more than one attribute key that contains that phrase in a single record, you can use the limit function to just find the first hit. You could do something like build an array from all the key hits but that is more involved and not necessary in this example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5b9167dc",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 2/2 [00:00<00:00, 1462.70query/s]                        \n",
      "Downloading: 100%|██████████| 10571/10571 [00:01<00:00, 5628.02rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>source_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>breast_cancer_attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR015448</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>112586635</td>\n",
       "      <td>primary_search</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR5768015</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>other</td>\n",
       "      <td>PRJNA392337</td>\n",
       "      <td>MDA-MB-231</td>\n",
       "      <td>481000108</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR5768014</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>other</td>\n",
       "      <td>PRJNA392337</td>\n",
       "      <td>MDA-MB-231</td>\n",
       "      <td>461470848</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR12410836</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>other</td>\n",
       "      <td>PRJNA655843</td>\n",
       "      <td>Pool of 5 breast cancer cell lines</td>\n",
       "      <td>38765680</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR12409553</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>other</td>\n",
       "      <td>PRJNA655843</td>\n",
       "      <td>Pool of 5 breast cancer cell lines</td>\n",
       "      <td>50943462</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10566</th>\n",
       "      <td>SRR6464399</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>PRJNA429642</td>\n",
       "      <td>ERa E2 ChIP-seq</td>\n",
       "      <td>2327280400</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10567</th>\n",
       "      <td>SRR10852376</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>PRJNA599998</td>\n",
       "      <td>DCIScom-NS</td>\n",
       "      <td>527684948</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10568</th>\n",
       "      <td>SRR10961905</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>PRJNA603080</td>\n",
       "      <td>BT-474 human breast cancer cell line treated w...</td>\n",
       "      <td>1047737700</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10569</th>\n",
       "      <td>SRR1021790</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>PRJNA147213</td>\n",
       "      <td>Human breast cancer cell line</td>\n",
       "      <td>851894748</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10570</th>\n",
       "      <td>SRR2176968</td>\n",
       "      <td>ChIP-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>GENOMIC</td>\n",
       "      <td>ChIP</td>\n",
       "      <td>PRJNA293471</td>\n",
       "      <td>MCF-7 breast cancer cell line</td>\n",
       "      <td>1797114788</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10571 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               acc assay_type consent librarysource libraryselection  \\\n",
       "0        SRR015448      OTHER  public       GENOMIC           RT-PCR   \n",
       "1       SRR5768015      OTHER  public       GENOMIC            other   \n",
       "2       SRR5768014      OTHER  public       GENOMIC            other   \n",
       "3      SRR12410836      OTHER  public       GENOMIC            other   \n",
       "4      SRR12409553      OTHER  public       GENOMIC            other   \n",
       "...            ...        ...     ...           ...              ...   \n",
       "10566   SRR6464399   ChIP-Seq  public       GENOMIC             ChIP   \n",
       "10567  SRR10852376   ChIP-Seq  public       GENOMIC             ChIP   \n",
       "10568  SRR10961905   ChIP-Seq  public       GENOMIC             ChIP   \n",
       "10569   SRR1021790   ChIP-Seq  public       GENOMIC             ChIP   \n",
       "10570   SRR2176968   ChIP-Seq  public       GENOMIC             ChIP   \n",
       "\n",
       "        bioproject                                        source_name  \\\n",
       "0             None                                               None   \n",
       "1      PRJNA392337                                         MDA-MB-231   \n",
       "2      PRJNA392337                                         MDA-MB-231   \n",
       "3      PRJNA655843                 Pool of 5 breast cancer cell lines   \n",
       "4      PRJNA655843                 Pool of 5 breast cancer cell lines   \n",
       "...            ...                                                ...   \n",
       "10566  PRJNA429642                                    ERa E2 ChIP-seq   \n",
       "10567  PRJNA599998                                         DCIScom-NS   \n",
       "10568  PRJNA603080  BT-474 human breast cancer cell line treated w...   \n",
       "10569  PRJNA147213                      Human breast cancer cell line   \n",
       "10570  PRJNA293471                      MCF-7 breast cancer cell line   \n",
       "\n",
       "            bases breast_cancer_attribute  \n",
       "0       112586635          primary_search  \n",
       "1       481000108           cell_type_sam  \n",
       "2       461470848           cell_type_sam  \n",
       "3        38765680         source_name_sam  \n",
       "4        50943462         source_name_sam  \n",
       "...           ...                     ...  \n",
       "10566  2327280400           cell_type_sam  \n",
       "10567   527684948           cell_type_sam  \n",
       "10568  1047737700         source_name_sam  \n",
       "10569   851894748         source_name_sam  \n",
       "10570  1797114788         source_name_sam  \n",
       "\n",
       "[10571 rows x 9 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "    (select v from unnest(attributes) where k = 'source_name_sam')as source_name,\n",
    "    (select v from unnest(attributes) where k = 'bases')as bases,\n",
    "    (select k from unnest(attributes) where v like '%breast%cancer%cell%line%' limit 1) as breast_cancer_attribute,\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE extracted.v LIKE '%breast%cancer%cell%line%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40bca2c1",
   "metadata": {},
   "source": [
    "## Limiting the Search Further\n",
    "Now you will add the search about the Estrogen Receptor status of the cells as well as limit it to paired RNA-Seq datasets using cDNA selection that are in the public domain. And again you will query the keys to see what attribute contained the information about the cells being ER-positive.\n",
    "# paired RNA-Seq datasets using cDNA selection that are in the public domain and"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b04eac0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 639.47query/s]                          \n",
      "Downloading: 100%|██████████| 6/6 [00:00<00:00,  7.65rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>source_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>ER_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8495050</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7551937500</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8495052</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>7745716800</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8495049</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7359307500</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8495053</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>7849955700</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8495054</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>6637409100</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR8495051</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7402524600</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc assay_type consent   librarysource libraryselection  \\\n",
       "0  SRR8495050    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1  SRR8495052    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "2  SRR8495049    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "3  SRR8495053    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "4  SRR8495054    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "5  SRR8495051    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "\n",
       "    bioproject source_name       bases      ER_column  \n",
       "0  PRJNA517424        T47D  7551937500  cell_type_sam  \n",
       "1  PRJNA517424     T47D-TR  7745716800  cell_type_sam  \n",
       "2  PRJNA517424        T47D  7359307500  cell_type_sam  \n",
       "3  PRJNA517424     T47D-TR  7849955700  cell_type_sam  \n",
       "4  PRJNA517424     T47D-TR  6637409100  cell_type_sam  \n",
       "5  PRJNA517424        T47D  7402524600  cell_type_sam  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "    (select v from unnest(attributes) where k = 'source_name_sam')as source_name,\n",
    "    (select v from unnest(attributes) where k = 'bases')as bases,\n",
    "    (select k from unnest(attributes) where v like '%ER-positive%' limit 1) as ER_column\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'\n",
    "    AND extracted.v LIKE '%ER-positive%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e52a1e3",
   "metadata": {},
   "source": [
    "## Storing Accessions in a Variable\n",
    "In the last couple of examples we will look at getting a list of accessions as a file within Jupyter. This is a frequent need when working with data that is uniquely identified by accessions or UUIDs. You can store the results of the query in a variable using a function in the bigquery module. You do this by providing the name of the variable \"results\" to store the output in. This is somewhat specific to the bigquery python module you're using in Jupyter and might not apply to how you will do queries normally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "063c0d33",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 801.66query/s]                          \n",
      "Downloading: 100%|██████████| 6/6 [00:00<00:00,  8.05rows/s]\n"
     ]
    }
   ],
   "source": [
    "%%bigquery results\n",
    "\n",
    "SELECT acc\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'\n",
    "    AND extracted.v LIKE '%ER-positive%'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6e699b0",
   "metadata": {},
   "source": [
    "## Viewing the Contents of the Variable\n",
    "You can output the contents of a variable by running a cell with the variable name in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1a7d3e29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8495050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8495052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8495049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8495054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8495053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR8495051</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc\n",
       "0  SRR8495050\n",
       "1  SRR8495052\n",
       "2  SRR8495049\n",
       "3  SRR8495054\n",
       "4  SRR8495053\n",
       "5  SRR8495051"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f9fcbbf",
   "metadata": {},
   "source": [
    "## Storing the Variable Contents as an Accession List\n",
    "The results variable contains a dataframe with an index (0-5) and a header (acc) that you don't want in the accession list. You will remove those using the to_string function using the options of index=False and header=False. This will save a file with just a list of accessions. Alternatively if you wanted to save the results of a metadata query that included the header information, you can omit the header=False option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "6a25ab01",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open(\"accessions.txt\", \"w\")\n",
    "file.write(results.to_string(index=False, header=False))\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
