{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e2af07ff",
   "metadata": {},
   "source": [
    "# Getting Started with Arrays in BigQuery\n",
    "Arrays are an important topic to understand on at least a basic level to work with data in BigQuery. It allows additional information to be stored in lists or key-value pairs within the table. Both can be searched on and extracted to find exactly what you are looking for.\n",
    "\n",
    "First you need to install and prepare the environment again."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee747a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# These commands were run when the notebook started but are shown in case you choose to work with Notebooks later.\n",
    "!pip install google.cloud.bigquery\n",
    "!pip install google.cloud.storage\n",
    "!pip3 install --upgrade google-cloud-bigquery\n",
    "!pip install google --user\n",
    "!pip install --upgrade 'google-cloud-bigquery[bqstorage,pandas]' --user"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4b5fe4be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run this command block to install the BigQuery extension.\n",
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c5f6779",
   "metadata": {},
   "source": [
    "## What Does an Array look like in BigQuery?\n",
    "Now you can look at a couple of columns from the earlier searches to show examples of arrays. You will select the accession column and the columns that are either simple arrays or are an array of data structures. This is the original query for 5 human runs looking at just those columns with an array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18743fb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 2/2 [00:00<00:00, 1410.56query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  7.33rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DRR245014</td>\n",
       "      <td>[sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '1552245200'}, {'k': 'byt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR2535945</td>\n",
       "      <td>[sra, fastq]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '970498698'}, {'k': 'byte...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR15402535</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'not applicable'}, {'k...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR15611200</td>\n",
       "      <td>[bam, sra]</td>\n",
       "      <td>[ncbi]</td>\n",
       "      <td>[ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'base...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR15584106</td>\n",
       "      <td>[bam, sra]</td>\n",
       "      <td>[ncbi]</td>\n",
       "      <td>[ncbi.public]</td>\n",
       "      <td>[{'k': 'sex_calc', 'v': 'female'}, {'k': 'base...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc datastore_filetype datastore_provider  \\\n",
       "0    DRR245014              [sra]     [gs, ncbi, s3]   \n",
       "1   ERR2535945       [sra, fastq]     [gs, ncbi, s3]   \n",
       "2  SRR15402535       [fastq, sra]     [gs, ncbi, s3]   \n",
       "3  SRR15611200         [bam, sra]             [ncbi]   \n",
       "4  SRR15584106         [bam, sra]             [ncbi]   \n",
       "\n",
       "                     datastore_region  \\\n",
       "0  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "1  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "2  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "3                       [ncbi.public]   \n",
       "4                       [ncbi.public]   \n",
       "\n",
       "                                          attributes  \n",
       "0  [{'k': 'bases', 'v': '1552245200'}, {'k': 'byt...  \n",
       "1  [{'k': 'bases', 'v': '970498698'}, {'k': 'byte...  \n",
       "2  [{'k': 'sex_calc', 'v': 'not applicable'}, {'k...  \n",
       "3  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'base...  \n",
       "4  [{'k': 'sex_calc', 'v': 'female'}, {'k': 'base...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, datastore_filetype, datastore_provider, datastore_region, attributes\n",
    "FROM `nih-sra-datastore.sra.metadata`\n",
    "WHERE organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6c50bee",
   "metadata": {},
   "source": [
    "## Arrays Compared to Structs\n",
    "All of the columns other thatn **acc** are arrays \\[ \\] with a comma separated list of items inside. The **attributes** column is an array of structs { } that have a key (k) and a value (v) for each item in the array. Arrays with just a list of strings or integers are usually easy to read. You can use the UNNEST function in BigQuery to extract the contents of an array into one item per row. \n",
    "\n",
    "### Naming Unnested Data\n",
    "When you unnest the data you can give the result of the unnest function a name (data_host in this query) and then use that name in the select statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "cc404824",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 997.69query/s]                          \n",
      "Downloading: 100%|██████████| 3/3 [00:00<00:00,  4.79rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>data_host</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ncbi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>s3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  data_host\n",
       "0        gs\n",
       "1      ncbi\n",
       "2        s3"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT data_host\n",
    "FROM `nih-sra-datastore.sra.metadata`,\n",
    "    UNNEST(datastore_provider) as data_host\n",
    "WHERE acc = 'SRR2973262'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9692198c",
   "metadata": {},
   "source": [
    "## Listing the Contents of Structs\n",
    "If you unnest an array of structs, you get the list of the structs in the array. The structure of the column \"attributes\" for SRA is a list of key-value pairs. The contents of this column are some standard items like \"bytes\" and \"bases\" as well as items that were provided by the submitter as additional information. This means that not all runs in the database will have the same attributes. There can also be keys that are repeated so one key might have multiple values in a single run. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "713e06d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 584.49query/s]                          \n",
      "Downloading: 100%|██████████| 18/18 [00:00<00:00, 27.48rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>extracted_attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>{'k': 'sex_calc', 'v': 'female'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>{'k': 'bases', 'v': '12470223358'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>{'k': 'bytes', 'v': '6566142480'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>{'k': 'age_sam', 'v': 'NA'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>{'k': 'biomaterial_provider_sam', 'v': 'Robert...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>{'k': 'cell_line_sam_ss_dpl110', 'v': 'HCA-7'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>{'k': 'isolate_sam', 'v': 'cell lines'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>{'k': 'tissue_sam', 'v': 'colon'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>{'k': 'primary_search', 'v': '304768'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>{'k': 'primary_search', 'v': '4309137'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'HCA-7'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'HCA-7-fastq'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'PRJNA304768'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SAMN04309137'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRP066923'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRR2973262'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRS1187131'}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>{'k': 'primary_search', 'v': 'SRX1459675'}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 extracted_attributes\n",
       "0                    {'k': 'sex_calc', 'v': 'female'}\n",
       "1                  {'k': 'bases', 'v': '12470223358'}\n",
       "2                   {'k': 'bytes', 'v': '6566142480'}\n",
       "3                         {'k': 'age_sam', 'v': 'NA'}\n",
       "4   {'k': 'biomaterial_provider_sam', 'v': 'Robert...\n",
       "5      {'k': 'cell_line_sam_ss_dpl110', 'v': 'HCA-7'}\n",
       "6             {'k': 'isolate_sam', 'v': 'cell lines'}\n",
       "7                   {'k': 'tissue_sam', 'v': 'colon'}\n",
       "8              {'k': 'primary_search', 'v': '304768'}\n",
       "9             {'k': 'primary_search', 'v': '4309137'}\n",
       "10              {'k': 'primary_search', 'v': 'HCA-7'}\n",
       "11        {'k': 'primary_search', 'v': 'HCA-7-fastq'}\n",
       "12        {'k': 'primary_search', 'v': 'PRJNA304768'}\n",
       "13       {'k': 'primary_search', 'v': 'SAMN04309137'}\n",
       "14          {'k': 'primary_search', 'v': 'SRP066923'}\n",
       "15         {'k': 'primary_search', 'v': 'SRR2973262'}\n",
       "16         {'k': 'primary_search', 'v': 'SRS1187131'}\n",
       "17         {'k': 'primary_search', 'v': 'SRX1459675'}"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT unnested_attributes\n",
    "FROM `nih-sra-datastore.sra.metadata`,\n",
    "    UNNEST(attributes) as unnested_attributes \n",
    "WHERE acc = 'SRR2973262'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9962febe",
   "metadata": {},
   "source": [
    "## Listing the Keys and Values in the Attributes Structs as a Table\n",
    "You can further break a struct list like this {'k': 'sex_calc', 'v': 'female'} apart into (k) and (v) columns by using unnested_attributes.k and unnested_attributes.v in our query. At this point you will be seeing all the keys and valunes in a table format for a single accession in the metadata table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d2ec3b3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 791.68query/s]                          \n",
      "Downloading: 100%|██████████| 18/18 [00:00<00:00, 26.09rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k</th>\n",
       "      <th>v</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>sex_calc</td>\n",
       "      <td>female</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>bases</td>\n",
       "      <td>12470223358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>bytes</td>\n",
       "      <td>6566142480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>age_sam</td>\n",
       "      <td>NA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "      <td>Robert Coffey, Vanderbilt University</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>cell_line_sam_ss_dpl110</td>\n",
       "      <td>HCA-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>isolate_sam</td>\n",
       "      <td>cell lines</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tissue_sam</td>\n",
       "      <td>colon</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>304768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>4309137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>HCA-7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>HCA-7-fastq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>PRJNA304768</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SAMN04309137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRP066923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRR2973262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRS1187131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>primary_search</td>\n",
       "      <td>SRX1459675</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           k                                     v\n",
       "0                   sex_calc                                female\n",
       "1                      bases                           12470223358\n",
       "2                      bytes                            6566142480\n",
       "3                    age_sam                                    NA\n",
       "4   biomaterial_provider_sam  Robert Coffey, Vanderbilt University\n",
       "5    cell_line_sam_ss_dpl110                                 HCA-7\n",
       "6                isolate_sam                            cell lines\n",
       "7                 tissue_sam                                 colon\n",
       "8             primary_search                                304768\n",
       "9             primary_search                               4309137\n",
       "10            primary_search                                 HCA-7\n",
       "11            primary_search                           HCA-7-fastq\n",
       "12            primary_search                           PRJNA304768\n",
       "13            primary_search                          SAMN04309137\n",
       "14            primary_search                             SRP066923\n",
       "15            primary_search                            SRR2973262\n",
       "16            primary_search                            SRS1187131\n",
       "17            primary_search                            SRX1459675"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT unnested_attributes.k, unnested_attributes.v\n",
    "FROM `nih-sra-datastore.sra.metadata`,\n",
    "    UNNEST(attributes) as unnested_attributes \n",
    "WHERE acc = 'SRR2973262'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb2de6d",
   "metadata": {},
   "source": [
    "## Finding Runs with a Certain Attribute\n",
    "Listing out all the attribute keys and values can be quite useful to help build searches for very specific things. Looking at the results for the above listing, one of the attributes is tissue_sam : colon. If you want to find all runs where that exists you can use the following query."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "b31b4101",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 946.58query/s]                          \n",
      "Downloading: 100%|██████████| 15550/15550 [00:01<00:00, 13305.05rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR14684549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR15652167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR7762696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR7762697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR7762698</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15545</th>\n",
       "      <td>SRR13345069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15546</th>\n",
       "      <td>SRR13872307</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15547</th>\n",
       "      <td>SRR15347514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15548</th>\n",
       "      <td>SRR11631045</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15549</th>\n",
       "      <td>SRR5047917</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15550 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               acc\n",
       "0      SRR14684549\n",
       "1      SRR15652167\n",
       "2       SRR7762696\n",
       "3       SRR7762697\n",
       "4       SRR7762698\n",
       "...            ...\n",
       "15545  SRR13345069\n",
       "15546  SRR13872307\n",
       "15547  SRR15347514\n",
       "15548  SRR11631045\n",
       "15549   SRR5047917\n",
       "\n",
       "[15550 rows x 1 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc\n",
    "FROM `nih-sra-datastore.sra.metadata`   \n",
    "WHERE ('tissue_sam', 'colon') in UNNEST(attributes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca8d90c1",
   "metadata": {},
   "source": [
    "## Search for Runs to Output Data\n",
    "Before moving on to using the SRA Toolkit you need to get a list of accessions to work with. One very useful part of Entrez on the NCBI website is natural language searching. You can type \"breast cancer cell line\" into the search and get a large number of results that probaby contain what you are looking for. You can replicate some of that search by using the LIKE function with the wildcard % in our attributes search. The query below will search for all paired RNA-Seq datasets using cDNA selection that are in the public domain and mention breast cancer cell line in the attributes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81aa0231",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 1194.28query/s]                         \n",
      "Downloading: 100%|██████████| 1894/1894 [00:01<00:00, 1688.42rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR9134779</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR9134779</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR9134766</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR9134766</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR9134783</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1889</th>\n",
       "      <td>SRR9077653</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA543423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1890</th>\n",
       "      <td>SRR9077653</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA543423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>SRR9134489</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1892</th>\n",
       "      <td>SRR9134489</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>SRR14368528</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA726397</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1894 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              acc assay_type consent   librarysource libraryselection  \\\n",
       "0      SRR9134779    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1      SRR9134779    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "2      SRR9134766    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "3      SRR9134766    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "4      SRR9134783    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "...           ...        ...     ...             ...              ...   \n",
       "1889   SRR9077653    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1890   SRR9077653    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1891   SRR9134489    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1892   SRR9134489    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1893  SRR14368528    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "\n",
       "       bioproject  \n",
       "0     PRJNA504037  \n",
       "1     PRJNA504037  \n",
       "2     PRJNA504037  \n",
       "3     PRJNA504037  \n",
       "4     PRJNA504037  \n",
       "...           ...  \n",
       "1889  PRJNA543423  \n",
       "1890  PRJNA543423  \n",
       "1891  PRJNA504037  \n",
       "1892  PRJNA504037  \n",
       "1893  PRJNA726397  \n",
       "\n",
       "[1894 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a26bf4aa",
   "metadata": {},
   "source": [
    "### Subselect\n",
    "A subselect is a select statement embedded within another select statement. In this case you will use a subselect to get the value of attributes using the attribute key name. You will use 'source_name_sam' and 'bases' as an example because they are likely to be on most records.\n",
    "\n",
    "You can also find the key for the attribute that contained the breast cancer cell line entry by using a subselect. Because there might be more than one attribute key that contains that phrase in a single record, you can use the limit function to just find the first hit. You could do something like build an array from all the key hits, but that is more involved and not necessary in this example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9bfac615",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 1025.00query/s]                         \n",
      "Downloading: 100%|██████████| 1894/1894 [00:01<00:00, 1484.61rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>source_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>breast_cancer_attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR9134779</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "      <td>None</td>\n",
       "      <td>262532850</td>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR9134779</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "      <td>None</td>\n",
       "      <td>262532850</td>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR9134766</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "      <td>None</td>\n",
       "      <td>5640579330</td>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR9134766</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "      <td>None</td>\n",
       "      <td>5640579330</td>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR9134783</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA504037</td>\n",
       "      <td>None</td>\n",
       "      <td>6792583094</td>\n",
       "      <td>biomaterial_provider_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1889</th>\n",
       "      <td>SRR13815301</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA705671</td>\n",
       "      <td>SUM159_HIF-1a and HIF-2a double knockdown_1% O2</td>\n",
       "      <td>17699568284</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1890</th>\n",
       "      <td>SRR13239745</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA684349</td>\n",
       "      <td>MCF7 breast cancer cell line</td>\n",
       "      <td>8275186258</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1891</th>\n",
       "      <td>SRR10916796</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA602306</td>\n",
       "      <td>breast cancer cell line</td>\n",
       "      <td>3965484990</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1892</th>\n",
       "      <td>SRR925704</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA210428</td>\n",
       "      <td>breast cancer cell line</td>\n",
       "      <td>4835439048</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1893</th>\n",
       "      <td>SRR925720</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA210428</td>\n",
       "      <td>breast cancer cell line</td>\n",
       "      <td>4848508008</td>\n",
       "      <td>source_name_sam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1894 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              acc assay_type consent   librarysource libraryselection  \\\n",
       "0      SRR9134779    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1      SRR9134779    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "2      SRR9134766    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "3      SRR9134766    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "4      SRR9134783    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "...           ...        ...     ...             ...              ...   \n",
       "1889  SRR13815301    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1890  SRR13239745    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1891  SRR10916796    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1892    SRR925704    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1893    SRR925720    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "\n",
       "       bioproject                                      source_name  \\\n",
       "0     PRJNA504037                                             None   \n",
       "1     PRJNA504037                                             None   \n",
       "2     PRJNA504037                                             None   \n",
       "3     PRJNA504037                                             None   \n",
       "4     PRJNA504037                                             None   \n",
       "...           ...                                              ...   \n",
       "1889  PRJNA705671  SUM159_HIF-1a and HIF-2a double knockdown_1% O2   \n",
       "1890  PRJNA684349                     MCF7 breast cancer cell line   \n",
       "1891  PRJNA602306                          breast cancer cell line   \n",
       "1892  PRJNA210428                          breast cancer cell line   \n",
       "1893  PRJNA210428                          breast cancer cell line   \n",
       "\n",
       "            bases   breast_cancer_attribute  \n",
       "0       262532850  biomaterial_provider_sam  \n",
       "1       262532850  biomaterial_provider_sam  \n",
       "2      5640579330  biomaterial_provider_sam  \n",
       "3      5640579330  biomaterial_provider_sam  \n",
       "4      6792583094  biomaterial_provider_sam  \n",
       "...           ...                       ...  \n",
       "1889  17699568284             cell_type_sam  \n",
       "1890   8275186258           source_name_sam  \n",
       "1891   3965484990           source_name_sam  \n",
       "1892   4835439048           source_name_sam  \n",
       "1893   4848508008           source_name_sam  \n",
       "\n",
       "[1894 rows x 9 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "    (select v from unnest(attributes) where k = 'source_name_sam')as source_name,\n",
    "    (select v from unnest(attributes) where k = 'bases')as bases,\n",
    "    (select k from unnest(attributes) where v like '%breast%cancer%cell%line%' limit 1) as breast_cancer_attribute,\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4c95b13",
   "metadata": {},
   "source": [
    "## Limiting the Search Further\n",
    "There are still quite a few records in this list so you might be able to find an even more specific attribute. Now you will add a search about the Estrogen Receptor status of cells. And again you will query the keys to see what attribute contained the information about the cells being ER-positive."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b04eac0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 750.32query/s]                          \n",
      "Downloading: 100%|██████████| 6/6 [00:00<00:00,  9.90rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>consent</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>source_name</th>\n",
       "      <th>bases</th>\n",
       "      <th>ER_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8495053</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>7849955700</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8495049</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7359307500</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8495051</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7402524600</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8495054</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>6637409100</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8495050</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D</td>\n",
       "      <td>7551937500</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR8495052</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>public</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>PRJNA517424</td>\n",
       "      <td>T47D-TR</td>\n",
       "      <td>7745716800</td>\n",
       "      <td>cell_type_sam</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc assay_type consent   librarysource libraryselection  \\\n",
       "0  SRR8495053    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "1  SRR8495049    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "2  SRR8495051    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "3  SRR8495054    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "4  SRR8495050    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "5  SRR8495052    RNA-Seq  public  TRANSCRIPTOMIC             cDNA   \n",
       "\n",
       "    bioproject source_name       bases      ER_column  \n",
       "0  PRJNA517424     T47D-TR  7849955700  cell_type_sam  \n",
       "1  PRJNA517424        T47D  7359307500  cell_type_sam  \n",
       "2  PRJNA517424        T47D  7402524600  cell_type_sam  \n",
       "3  PRJNA517424     T47D-TR  6637409100  cell_type_sam  \n",
       "4  PRJNA517424        T47D  7551937500  cell_type_sam  \n",
       "5  PRJNA517424     T47D-TR  7745716800  cell_type_sam  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "\n",
    "SELECT acc, assay_type, consent, librarysource, libraryselection, bioproject,\n",
    "    (select v from unnest(attributes) where k = 'source_name_sam')as source_name,\n",
    "    (select v from unnest(attributes) where k = 'bases')as bases,\n",
    "    (select k from unnest(attributes) where v like '%ER-positive%' limit 1) as ER_column\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'\n",
    "    AND extracted.v LIKE '%ER-positive%'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e52a1e3",
   "metadata": {},
   "source": [
    "## Storing Accessions in a Variable\n",
    "You can store the results of the query in a variable using a function in the bigquery module. You do this by providing the name of the variable \"results\" to store the output in. This is somewhat specific to the bigquery python module you're using in Jupyter and might not apply to how you will do queries normally."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "063c0d33",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bigquery results\n",
    "\n",
    "SELECT acc\n",
    "FROM `nih-sra-datastore.sra.metadata` meta,\n",
    "    UNNEST(attributes) as extracted\n",
    "WHERE assay_type = 'RNA-Seq'\n",
    "    AND consent = 'public'\n",
    "    AND libraryselection = 'cDNA'\n",
    "    AND librarylayout = 'PAIRED'\n",
    "    AND extracted.v LIKE '%breast%cancer%cell%line%'\n",
    "    AND extracted.v LIKE '%ER-positive%'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d6e699b0",
   "metadata": {},
   "source": [
    "## Viewing the Contents of the Variable\n",
    "You can output the contents of a variable by running a cell with the variable name in it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "1a7d3e29",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR8495049</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR8495052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR8495053</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR8495050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR8495051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SRR8495054</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc\n",
       "0  SRR8495049\n",
       "1  SRR8495052\n",
       "2  SRR8495053\n",
       "3  SRR8495050\n",
       "4  SRR8495051\n",
       "5  SRR8495054"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f9fcbbf",
   "metadata": {},
   "source": [
    "## Storing the Variable Contents as an Accession List\n",
    "The results variable contains a datafram with an index (0-5) and a header (acc) that you don't want in the accession list. You will remove those using the to_string function using the options of index=False and header=False. This will save a file with just a list of accessions. Alternatively if you wanted to save the results of a metadata query that included the header information, you can omit the header=False option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "6a25ab01",
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open(\"accessions.txt\", \"w\")\n",
    "file.write(results.to_string(index=False, header=False))\n",
    "file.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
