{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fbb8205",
   "metadata": {},
   "source": [
    "# Getting Started with BigQuery SQL Searches\n",
    "First we need to install some packages and load the BigQuery extension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f34a6648",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting google.cloud.bigquery\n",
      "  Using cached https://files.pythonhosted.org/packages/fd/2a/d53b342d20e4b95ade480fa04977969166bb189bf0e909501e4fff86cb3f/google_cloud_bigquery-2.25.1-py2.py3-none-any.whl\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (2.22.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (3.17.3)\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.19.0)\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.7.2)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.39.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (21.0)\n",
      "Requirement already satisfied: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.31.1)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.3.3)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (1.24.2)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2021.5.30)\n",
      "Requirement already satisfied: six>=1.9 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from protobuf>=3.12.0->google.cloud.bigquery) (1.16.0)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (1.34.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google.cloud.bigquery) (2.4.7)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (2021.1)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (41.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (1.53.0)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.1.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.7.2)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.2.2)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.12.3)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.4.8)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (2.19)\n",
      "Installing collected packages: google.cloud.bigquery\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_bigquery-2.25.1-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google.cloud.storage\n",
      "  Using cached https://files.pythonhosted.org/packages/0e/d6/5878d73105fd242dafb42bbea26629372d397f06cb402e90302a4824c2c2/google_cloud_storage-1.42.0-py2.py3-none-any.whl\n",
      "Requirement already satisfied: google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.31.1)\n",
      "Requirement already satisfied: google-cloud-core<3.0dev,>=1.6.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.7.2)\n",
      "Requirement already satisfied: google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.34.0)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.3.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.storage) (2.22.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (3.17.3)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (21.0)\n",
      "Requirement already satisfied: six>=1.13.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (1.16.0)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (1.53.0)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2021.1)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (41.0.1)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.2.8)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.7.2)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.2.2)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.1.2)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2021.5.30)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2.8)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (1.24.2)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2.4.7)\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.4.8)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.12.3)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (2.19)\n",
      "Installing collected packages: google.cloud.storage\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_storage-1.42.0-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google-cloud-bigquery\n",
      "\u001b[33m  Cache entry deserialization failed, entry ignored\u001b[0m\n",
      "  Using cached https://files.pythonhosted.org/packages/fd/2a/d53b342d20e4b95ade480fa04977969166bb189bf0e909501e4fff86cb3f/google_cloud_bigquery-2.25.1-py2.py3-none-any.whl\n",
      "Collecting packaging>=14.3 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/3c/77/e2362b676dc5008d81be423070dd9577fa03be5da2ba1105811900fda546/packaging-21.0-py3-none-any.whl\n",
      "Collecting protobuf>=3.12.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/53/4e/e2db88d0bb0bda6a879eea62fddbaf813719ce3770d458bc5580512d9c95/protobuf-3.17.3-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.whl\n",
      "Collecting google-resumable-media<3.0dev,>=0.6.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/2b/86/ff9b23ac7446ee224d0d7e5e64efc433aeef52b76662863873e03259ed0f/google_resumable_media-2.0.0-py2.py3-none-any.whl\n",
      "Collecting google-api-core[grpc]<3.0.0dev,>=1.29.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/b4/5e/19075a55f9a44ac665cb3e9261622ca278c58252eba64d9c3f334245c00b/google_api_core-2.0.0-py2.py3-none-any.whl\n",
      "Collecting grpcio<2.0dev,>=1.38.1 (from google-cloud-bigquery)\n",
      "Collecting requests<3.0.0dev,>=2.18.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/92/96/144f70b972a9c0eabbd4391ef93ccd49d0f2747f4f6a2a2738e99e5adc65/requests-2.26.0-py2.py3-none-any.whl\n",
      "Collecting proto-plus>=1.10.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8c/72/6f3f4cdc5bb0294f8d7f3f8aacb617b4c3cb17554ed78f7e28009162c795/proto_plus-1.19.0-py3-none-any.whl\n",
      "Collecting google-cloud-core<3.0.0dev,>=1.4.1 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/42/b8/4ad3e423105d26a0c5eea54320aed57b7be060d1cd9faea53da713b408c6/google_cloud_core-2.0.0-py2.py3-none-any.whl\n",
      "Collecting pyparsing>=2.0.2 (from packaging>=14.3->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8a/bb/488841f56197b13700afd5658fc279a2025a39e22449b7cf29864669b15d/pyparsing-2.4.7-py2.py3-none-any.whl\n",
      "Collecting six>=1.9 (from protobuf>=3.12.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl\n",
      "Collecting google-crc32c<2.0dev,>=1.0 (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/5c/81/b71c645bac55d5c8f2d6e90603007c2940521ed63b177b833d1cd3ba7df8/google_crc32c-1.1.2-cp36-cp36m-manylinux1_x86_64.whl\n",
      "Collecting google-auth<3.0dev,>=1.25.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/b0/e2/940b73a4923b8659103bbabbf35d4fdeb11374f6bd47234f724238a6023b/google_auth-2.0.1-py2.py3-none-any.whl\n",
      "Collecting setuptools>=40.3.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bd/25/5bdf7f1adeebd4e3fa76b2e2f045ae53ee208e40a4231ad0f0c3007e4353/setuptools-57.4.0-py3-none-any.whl\n",
      "Collecting googleapis-common-protos<2.0dev,>=1.6.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/55/08/796a6bc0b550e2b7116041c953d3d5100016abea106131d71e5651826e7b/googleapis_common_protos-1.53.0-py2.py3-none-any.whl\n",
      "Collecting urllib3<1.27,>=1.21.1 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/5f/64/43575537846896abac0b15c3e5ac678d787a4021e906703f1766bfb8ea11/urllib3-1.26.6-py2.py3-none-any.whl\n",
      "Collecting idna<4,>=2.5; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d7/77/ff688d1504cdc4db2a938e2b7b9adee5dd52e34efbd2431051efc9984de9/idna-3.2-py3-none-any.whl\n",
      "Collecting charset-normalizer~=2.0.0; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/33/53/b7f6126a2b9fd878b025fe3c40266cfaad696f312165008ce045bffa3fe7/charset_normalizer-2.0.4-py3-none-any.whl\n",
      "Collecting certifi>=2017.4.17 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/05/1b/0a0dece0e8aa492a6ec9e4ad2fe366b511558cdc73fd3abc82ba7348e875/certifi-2021.5.30-py2.py3-none-any.whl\n",
      "Collecting cffi>=1.0.0 (from google-crc32c<2.0dev,>=1.0->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/f2/cd/3f5f059fed635d71047fa9ce507635088f982ab280fc24cde91d9afb9c1c/cffi-1.14.6-cp36-cp36m-manylinux1_x86_64.whl\n",
      "Collecting cachetools<5.0,>=2.0.0 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bf/28/c4f5796c67ad06bb91d98d543a5e01805c1ff065e08871f78e52d2a331ad/cachetools-4.2.2-py3-none-any.whl\n",
      "Collecting rsa<5,>=3.1.4 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/e9/93/0c0f002031f18b53af7a6166103c02b9c0667be528944137cc954ec921b3/rsa-4.7.2-py3-none-any.whl\n",
      "Collecting pyasn1-modules>=0.2.1 (from google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/95/de/214830a981892a3e286c3794f41ae67a4495df1108c3da8a9f62159b9a9d/pyasn1_modules-0.2.8-py2.py3-none-any.whl\n",
      "Collecting pycparser (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/ae/e7/d9c3a176ca4b02024debf82342dab36efadfc5776f9c8db077e8f6e71821/pycparser-2.20-py2.py3-none-any.whl\n",
      "Collecting pyasn1>=0.1.3 (from rsa<5,>=3.1.4->google-auth<3.0dev,>=1.25.0->google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/62/1e/a94a8d635fa3ce4cfc7f506003548d0a2447ae76fd5ca53932970fe3053f/pyasn1-0.4.8-py2.py3-none-any.whl\n",
      "Installing collected packages: pyparsing, packaging, six, protobuf, pycparser, cffi, google-crc32c, google-resumable-media, cachetools, pyasn1, rsa, setuptools, pyasn1-modules, google-auth, urllib3, idna, charset-normalizer, certifi, requests, googleapis-common-protos, grpcio, google-api-core, proto-plus, google-cloud-core, google-cloud-bigquery\n",
      "Successfully installed cachetools-4.2.2 certifi-2021.5.30 cffi-1.14.6 charset-normalizer-2.0.4 google-api-core-2.0.0 google-auth-2.0.1 google-cloud-bigquery-2.25.1 google-cloud-core-2.0.0 google-crc32c-1.1.2 google-resumable-media-2.0.0 googleapis-common-protos-1.53.0 grpcio-1.39.0 idna-3.2 packaging-21.0 proto-plus-1.19.0 protobuf-3.17.3 pyasn1-0.4.8 pyasn1-modules-0.2.8 pycparser-2.20 pyparsing-2.4.7 requests-2.26.0 rsa-4.7.2 setuptools-57.4.0 six-1.16.0 urllib3-1.26.6\n",
      "Requirement already satisfied: google in /home/jupyter-astine/.local/lib/python3.7/site-packages (3.0.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google) (4.9.3)\n",
      "Requirement already satisfied: soupsieve>1.2; python_version >= \"3.0\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from beautifulsoup4->google) (2.2.1)\n",
      "Requirement already up-to-date: google-cloud-bigquery[bqstorage,pandas] in /home/jupyter-astine/.local/lib/python3.7/site-packages (2.25.1)\n",
      "Requirement already satisfied, skipping upgrade: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.39.0)\n",
      "Requirement already satisfied, skipping upgrade: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (21.0)\n",
      "Requirement already satisfied, skipping upgrade: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.22.0)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.7.2)\n",
      "Requirement already satisfied, skipping upgrade: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.3)\n",
      "Requirement already satisfied, skipping upgrade: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (3.17.3)\n",
      "Requirement already satisfied, skipping upgrade: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.19.0)\n",
      "Requirement already satisfied, skipping upgrade: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.31.1)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.6.3)\n",
      "Requirement already satisfied, skipping upgrade: pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (5.0.0)\n",
      "Requirement already satisfied, skipping upgrade: pandas>=0.23.0; extra == \"pandas\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.1)\n",
      "Requirement already satisfied, skipping upgrade: six>=1.5.2 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from grpcio<2.0dev,>=1.38.1->google-cloud-bigquery[bqstorage,pandas]) (1.16.0)\n",
      "Requirement already satisfied, skipping upgrade: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-cloud-bigquery[bqstorage,pandas]) (2.4.7)\n",
      "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (1.24.2)\n",
      "Requirement already satisfied, skipping upgrade: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (3.0.4)\n",
      "Requirement already satisfied, skipping upgrade: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2.8)\n",
      "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2021.5.30)\n",
      "Requirement already satisfied, skipping upgrade: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (1.34.0)\n",
      "Requirement already satisfied, skipping upgrade: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.1.2)\n",
      "Requirement already satisfied, skipping upgrade: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (2021.1)\n",
      "Requirement already satisfied, skipping upgrade: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (1.53.0)\n",
      "Requirement already satisfied, skipping upgrade: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (41.0.1)\n",
      "Requirement already satisfied, skipping upgrade: libcst>=0.2.5 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.3.20)\n",
      "Requirement already satisfied, skipping upgrade: numpy>=1.16.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (1.21.1)\n",
      "Requirement already satisfied, skipping upgrade: python-dateutil>=2.7.3 in /opt/tljh/user/lib/python3.7/site-packages (from pandas>=0.23.0; extra == \"pandas\"->google-cloud-bigquery[bqstorage,pandas]) (2.8.2)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.2.8)\n",
      "Requirement already satisfied, skipping upgrade: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.7.2)\n",
      "Requirement already satisfied, skipping upgrade: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.2.2)\n",
      "Requirement already satisfied, skipping upgrade: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.12.3)\n",
      "Requirement already satisfied, skipping upgrade: typing-extensions>=3.7.4.2 in /opt/tljh/user/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (3.10.0.0)\n",
      "Requirement already satisfied, skipping upgrade: pyyaml>=5.2 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (5.4.1)\n",
      "Requirement already satisfied, skipping upgrade: typing-inspect>=0.4.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.7.1)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1<0.5.0,>=0.4.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.4.8)\n",
      "Requirement already satisfied, skipping upgrade: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (2.19)\n",
      "Requirement already satisfied, skipping upgrade: mypy-extensions>=0.3.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from typing-inspect>=0.4.0->libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.4.3)\n"
     ]
    }
   ],
   "source": [
    "!pip install google.cloud.bigquery\n",
    "!pip install google.cloud.storage\n",
    "!pip3 install --upgrade google-cloud-bigquery\n",
    "\n",
    "!pip install google --user\n",
    "!pip install --upgrade 'google-cloud-bigquery[bqstorage,pandas]' --user\n",
    "\n",
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0c3c81",
   "metadata": {},
   "source": [
    "## Basic Query\n",
    "Here is a simple query to find 15 runs submitted for the organsim 'Homo sapiens' from the nih-sra-datastore project. This query searches for all columns in the metadata table from the sra database of the project. We will filter for only the rows where the organism name is 'Homo sapiens' and limit the output to only the first 5 rows."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81be3e67",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 2/2 [00:00<00:00, 1737.49query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  7.33rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_country_continent_calc</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR15278393</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>EMORY UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11583084</td>\n",
       "      <td>2815176</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '9518138700'}, {'k': 'byt...</td>\n",
       "      <td>{\"bases\": 9518138700, \"bytes\": 3239993193, \"ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR15278347</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>EMORY UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11583130</td>\n",
       "      <td>919115</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '9340662300'}, {'k': 'byt...</td>\n",
       "      <td>{\"bases\": 9340662300, \"bytes\": 2922928789, \"ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR15065199</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375204</td>\n",
       "      <td>GSM5412973</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '188792856'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 188792856, \"bytes\": 85176148, \"age_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR15065057</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375346</td>\n",
       "      <td>GSM5412928</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '409589460'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 409589460, \"bytes\": 186427640, \"age_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR15065033</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375370</td>\n",
       "      <td>GSM5412909</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '245428848'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 245428848, \"bytes\": 103593882, \"age_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc assay_type                    center_name consent   experiment  \\\n",
       "0  SRR15278393    RNA-Seq               EMORY UNIVERSITY  public  SRX11583084   \n",
       "1  SRR15278347    RNA-Seq               EMORY UNIVERSITY  public  SRX11583130   \n",
       "2  SRR15065199   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375204   \n",
       "3  SRR15065057   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375346   \n",
       "4  SRR15065033   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375370   \n",
       "\n",
       "  sample_name           instrument librarylayout libraryselection  \\\n",
       "0     2815176  Illumina HiSeq 1000        PAIRED            PolyA   \n",
       "1      919115  Illumina HiSeq 1000        PAIRED            PolyA   \n",
       "2  GSM5412973  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "3  GSM5412928  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "4  GSM5412909  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "\n",
       "    librarysource  ... geo_loc_name_country_continent_calc geo_loc_name_sam  \\\n",
       "0  TRANSCRIPTOMIC  ...                                None               []   \n",
       "1  TRANSCRIPTOMIC  ...                                None               []   \n",
       "2     METAGENOMIC  ...                                None               []   \n",
       "3     METAGENOMIC  ...                                None               []   \n",
       "4     METAGENOMIC  ...                                None               []   \n",
       "\n",
       "  ena_first_public_run ena_last_update_run sample_name_sam datastore_filetype  \\\n",
       "0                   []                  []              []       [fastq, sra]   \n",
       "1                   []                  []              []       [fastq, sra]   \n",
       "2                   []                  []              []       [fastq, sra]   \n",
       "3                   []                  []              []       [fastq, sra]   \n",
       "4                   []                  []              []       [fastq, sra]   \n",
       "\n",
       "  datastore_provider                    datastore_region  \\\n",
       "0     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "1     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "2     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "3     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "4     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "\n",
       "                                          attributes  \\\n",
       "0  [{'k': 'bases', 'v': '9518138700'}, {'k': 'byt...   \n",
       "1  [{'k': 'bases', 'v': '9340662300'}, {'k': 'byt...   \n",
       "2  [{'k': 'bases', 'v': '188792856'}, {'k': 'byte...   \n",
       "3  [{'k': 'bases', 'v': '409589460'}, {'k': 'byte...   \n",
       "4  [{'k': 'bases', 'v': '245428848'}, {'k': 'byte...   \n",
       "\n",
       "                                               jattr  \n",
       "0  {\"bases\": 9518138700, \"bytes\": 3239993193, \"ag...  \n",
       "1  {\"bases\": 9340662300, \"bytes\": 2922928789, \"ag...  \n",
       "2  {\"bases\": 188792856, \"bytes\": 85176148, \"age_s...  \n",
       "3  {\"bases\": 409589460, \"bytes\": 186427640, \"age_...  \n",
       "4  {\"bases\": 245428848, \"bytes\": 103593882, \"age_...  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT *\n",
    "FROM `nih-sra-datastore.sra.metadata`\n",
    "WHERE organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fea8d95",
   "metadata": {},
   "source": [
    "## Searching Other Tables\n",
    "The second database in the nih-sra-datastore project contains data from the SRA Taxonomy Analysis Tool (STAT) as well as the Taxonomy database. The analysis will use the Taxonomy Database Identifier (tax ID) in the results. You can find the tax_id for any entry in the taxonomy table using a search like the one below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "21842a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 945.94query/s]                          \n",
      "Downloading: 100%|██████████| 1/1 [00:00<00:00,  1.38rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>names</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>ileft</th>\n",
       "      <th>iright</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  parent_id     rank      sci_name  \\\n",
       "0    9606       9605  species  Homo sapiens   \n",
       "\n",
       "                                               names  ilevel   ileft  iright  \n",
       "0  [{'name': 'Home sapiens', 'name_class': 'missp...      30  212190  212191  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT * \n",
    "FROM `nih-sra-datastore.sra_tax_analysis_tool.taxonomy`\n",
    "WHERE sci_name = 'Homo sapiens'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f7938d",
   "metadata": {},
   "source": [
    "## Joining Two (or more) Tables\n",
    "Using a JOIN command will allow you to combine two different tables to find the information you are looking for. This allows data to be stored most efficiently in the database but also allows a user to view the information in a way they find easiest to understand. Below we'll simply combine the two queries from before. When specifying multiple tables to search we can also give an abbreviated name to the table ('m' and 't' in the query below) to make the rest of the query easier to type and read.\n",
    "When joining two or more tables we will need to tell the database which columns in the two tables are expected to align. We need a field that can be used to link each table. This value is what we are joining \"on\" in the statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "45a293a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 2786.91query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  5.96rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>names</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>ileft</th>\n",
       "      <th>iright</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR2511434</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530849</td>\n",
       "      <td>SAMEA1116052</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '222222604'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 222222604, \"bytes\": 188868157, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR2511621</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2531036</td>\n",
       "      <td>SAMEA1116239</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '171452718'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 171452718, \"bytes\": 135479064, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR2511146</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530561</td>\n",
       "      <td>SAMEA1115764</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '245005622'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 245005622, \"bytes\": 197800594, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR2511321</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530736</td>\n",
       "      <td>SAMEA1115939</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '150219600'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 150219600, \"bytes\": 121420950, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR2510985</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530400</td>\n",
       "      <td>SAMEA1103343</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '90789342'}, {'k': 'bytes...</td>\n",
       "      <td>{\"bases\": 90789342, \"bytes\": 73927119, \"alias_...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212190</td>\n",
       "      <td>212191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc assay_type                            center_name consent  \\\n",
       "0  ERR2511434   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "1  ERR2511621   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "2  ERR2511146   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "3  ERR2511321   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "4  ERR2510985   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "\n",
       "   experiment   sample_name instrument librarylayout libraryselection  \\\n",
       "0  ERX2530849  SAMEA1116052     Ion S5        SINGLE              PCR   \n",
       "1  ERX2531036  SAMEA1116239     Ion S5        SINGLE              PCR   \n",
       "2  ERX2530561  SAMEA1115764     Ion S5        SINGLE              PCR   \n",
       "3  ERX2530736  SAMEA1115939     Ion S5        SINGLE              PCR   \n",
       "4  ERX2530400  SAMEA1103343     Ion S5        SINGLE              PCR   \n",
       "\n",
       "  librarysource  ...                                         attributes  \\\n",
       "0   METAGENOMIC  ...  [{'k': 'bases', 'v': '222222604'}, {'k': 'byte...   \n",
       "1   METAGENOMIC  ...  [{'k': 'bases', 'v': '171452718'}, {'k': 'byte...   \n",
       "2   METAGENOMIC  ...  [{'k': 'bases', 'v': '245005622'}, {'k': 'byte...   \n",
       "3   METAGENOMIC  ...  [{'k': 'bases', 'v': '150219600'}, {'k': 'byte...   \n",
       "4   METAGENOMIC  ...  [{'k': 'bases', 'v': '90789342'}, {'k': 'bytes...   \n",
       "\n",
       "                                               jattr tax_id parent_id  \\\n",
       "0  {\"bases\": 222222604, \"bytes\": 188868157, \"alia...   9606      9605   \n",
       "1  {\"bases\": 171452718, \"bytes\": 135479064, \"alia...   9606      9605   \n",
       "2  {\"bases\": 245005622, \"bytes\": 197800594, \"alia...   9606      9605   \n",
       "3  {\"bases\": 150219600, \"bytes\": 121420950, \"alia...   9606      9605   \n",
       "4  {\"bases\": 90789342, \"bytes\": 73927119, \"alias_...   9606      9605   \n",
       "\n",
       "      rank      sci_name                                              names  \\\n",
       "0  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "1  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "2  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "3  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "4  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "\n",
       "   ilevel   ileft  iright  \n",
       "0      30  212190  212191  \n",
       "1      30  212190  212191  \n",
       "2      30  212190  212191  \n",
       "3      30  212190  212191  \n",
       "4      30  212190  212191  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT *\n",
    "FROM `nih-sra-datastore.sra.metadata` m \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.taxonomy` t \n",
    "    ON m.organism = t.sci_name\n",
    "WHERE t.tax_id = 9606\n",
    "    AND m.organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54f6a913",
   "metadata": {},
   "source": [
    "## Selecting Fewer Columns\n",
    "Using * in the select statement will show all the columns in the table. Often we don't need or want all of the columns so we can show only the columns we are interested in. We do this by naming the columns in the select statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "b89a5690",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 3543.23query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  6.69rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>instrument</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>sci_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR6178770</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>RT-PCR</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR6788172</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR5943843</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DRR238304</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>MinION</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR12961085</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>MinION</td>\n",
       "      <td>Inverse rRNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc assay_type instrument libraryselection   librarysource  \\\n",
       "0   SRR6178770    RNA-Seq     Ion S5           RT-PCR  TRANSCRIPTOMIC   \n",
       "1   SRR6788172    RNA-Seq     Ion S5            PolyA  TRANSCRIPTOMIC   \n",
       "2   SRR5943843    RNA-Seq     Ion S5             cDNA  TRANSCRIPTOMIC   \n",
       "3    DRR238304    RNA-Seq     MinION             cDNA  TRANSCRIPTOMIC   \n",
       "4  SRR12961085    RNA-Seq     MinION     Inverse rRNA  TRANSCRIPTOMIC   \n",
       "\n",
       "       sci_name  \n",
       "0  Homo sapiens  \n",
       "1  Homo sapiens  \n",
       "2  Homo sapiens  \n",
       "3  Homo sapiens  \n",
       "4  Homo sapiens  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT m.acc, m.assay_type, instrument, libraryselection, librarysource, sci_name\n",
    "FROM `nih-sra-datastore.sra.metadata` m \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.taxonomy` t \n",
    "    ON m.organism = t.sci_name\n",
    "WHERE t.tax_id = 9606\n",
    "    AND m.organism = 'Homo sapiens'\n",
    "    AND m.assay_type = 'RNA-Seq'\n",
    "    AND m.consent = 'public'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2b0ad1",
   "metadata": {},
   "source": [
    "## Using Where Clauses to Find Data of Interest\n",
    "Now let's look at an example for someone who wants to do an alignment. If you wanted to use the PGAP package from NCBI, you would want data that has a high percentage of only one species to generate an assembly and annotation for. We can use the output from STAT to do this. We can compare the self_count columns to the analysed spot count to find a hi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "969f780b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 2939.25query/s]                        \n",
      "Downloading: 100%|██████████| 16153/16153 [00:00<00:00, 16483.84rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>name</th>\n",
       "      <th>analyzed_spot_count</th>\n",
       "      <th>self_count</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1817002</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>608593</td>\n",
       "      <td>594332</td>\n",
       "      <td>0.976567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1816986</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>692180</td>\n",
       "      <td>673720</td>\n",
       "      <td>0.973331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1816981</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>630954</td>\n",
       "      <td>613122</td>\n",
       "      <td>0.971738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1816996</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>593342</td>\n",
       "      <td>575904</td>\n",
       "      <td>0.970611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1431120</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>3249346</td>\n",
       "      <td>3153217</td>\n",
       "      <td>0.970416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16148</th>\n",
       "      <td>SRR10665812</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>510003</td>\n",
       "      <td>459009</td>\n",
       "      <td>0.900012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16149</th>\n",
       "      <td>SRR2071980</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>1336913</td>\n",
       "      <td>1203238</td>\n",
       "      <td>0.900012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16150</th>\n",
       "      <td>SRR6038661</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>749339</td>\n",
       "      <td>674410</td>\n",
       "      <td>0.900007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16151</th>\n",
       "      <td>ERR1230393</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>803265</td>\n",
       "      <td>722940</td>\n",
       "      <td>0.900002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16152</th>\n",
       "      <td>SRR2061198</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>352405</td>\n",
       "      <td>317165</td>\n",
       "      <td>0.900001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16153 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               acc                    name  analyzed_spot_count  self_count  \\\n",
       "0       ERR1817002  Listeria monocytogenes               608593      594332   \n",
       "1       ERR1816986  Listeria monocytogenes               692180      673720   \n",
       "2       ERR1816981  Listeria monocytogenes               630954      613122   \n",
       "3       ERR1816996  Listeria monocytogenes               593342      575904   \n",
       "4       ERR1431120  Listeria monocytogenes              3249346     3153217   \n",
       "...            ...                     ...                  ...         ...   \n",
       "16148  SRR10665812  Listeria monocytogenes               510003      459009   \n",
       "16149   SRR2071980  Listeria monocytogenes              1336913     1203238   \n",
       "16150   SRR6038661  Listeria monocytogenes               749339      674410   \n",
       "16151   ERR1230393  Listeria monocytogenes               803265      722940   \n",
       "16152   SRR2061198  Listeria monocytogenes               352405      317165   \n",
       "\n",
       "       proportion  \n",
       "0        0.976567  \n",
       "1        0.973331  \n",
       "2        0.971738  \n",
       "3        0.970611  \n",
       "4        0.970416  \n",
       "...           ...  \n",
       "16148    0.900012  \n",
       "16149    0.900012  \n",
       "16150    0.900007  \n",
       "16151    0.900002  \n",
       "16152    0.900001  \n",
       "\n",
       "[16153 rows x 5 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT a.acc, a.name, i.analyzed_spot_count, a.self_count, a.self_count/i.analyzed_spot_count as proportion\n",
    "FROM `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis` a \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis_info` i \n",
    "    ON a.acc = i.acc\n",
    "WHERE a.tax_id = 1639\n",
    "    AND a.self_count/i.analyzed_spot_count > .9\n",
    "    AND i.analyzed_spot_count > 100000\n",
    "ORDER BY proportion DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "6748fd71",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 5/5 [00:00<00:00, 4783.65query/s]                        \n",
      "Downloading: 100%|██████████| 119/119 [00:00<00:00, 149.82rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>geo_loc_name_country_calc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>366077</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>196962</td>\n",
       "      <td>United Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17804</td>\n",
       "      <td>Ireland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15778</td>\n",
       "      <td>Spain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15653</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>Laos</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>1</td>\n",
       "      <td>Burundi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>1</td>\n",
       "      <td>Liberia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>Bulgaria</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>1</td>\n",
       "      <td>Bosnia and Herzegovina</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      count geo_loc_name_country_calc\n",
       "0    366077                       USA\n",
       "1    196962            United Kingdom\n",
       "2     17804                   Ireland\n",
       "3     15778                     Spain\n",
       "4     15653                 Australia\n",
       "..      ...                       ...\n",
       "114       1                      Laos\n",
       "115       1                   Burundi\n",
       "116       1                   Liberia\n",
       "117       1                  Bulgaria\n",
       "118       1    Bosnia and Herzegovina\n",
       "\n",
       "[119 rows x 2 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT count(m.acc) as count, m.geo_loc_name_country_calc\n",
    "FROM `nih-sra-datastore.sra.metadata` m , `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis` tax\n",
    "WHERE m.acc = tax.acc\n",
    "    and tax.name = 'Coronaviridae'\n",
    "    and m.geo_loc_name_country_calc is NOT NULL\n",
    "GROUP BY m.geo_loc_name_country_calc\n",
    "ORDER by 1 DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ee663f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
