{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4fbb8205",
   "metadata": {},
   "source": [
    "# Getting Started with BigQuery SQL Searches\n",
    "First we need to install some packages and load the BigQuery extension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f34a6648",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting google.cloud.bigquery\n",
      "  Using cached https://files.pythonhosted.org/packages/fd/2a/d53b342d20e4b95ade480fa04977969166bb189bf0e909501e4fff86cb3f/google_cloud_bigquery-2.25.1-py2.py3-none-any.whl\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.7.2)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (21.0)\n",
      "Requirement already satisfied: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.19.0)\n",
      "Requirement already satisfied: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.31.1)\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.39.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (3.17.3)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.bigquery) (1.3.3)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.bigquery) (2.22.0)\n",
      "Requirement already satisfied: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (1.34.0)\n",
      "Requirement already satisfied: six>=1.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (1.16.0)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google.cloud.bigquery) (2.4.7)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (41.0.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (1.53.0)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google.cloud.bigquery) (2021.1)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.1.2)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (1.25.11)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2.8)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.bigquery) (2021.5.30)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.2.2)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (4.7.2)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.2.8)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (1.12.3)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google.cloud.bigquery) (0.4.8)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google.cloud.bigquery) (2.19)\n",
      "Installing collected packages: google.cloud.bigquery\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_bigquery-2.25.1-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google.cloud.storage\n",
      "  Using cached https://files.pythonhosted.org/packages/0e/d6/5878d73105fd242dafb42bbea26629372d397f06cb402e90302a4824c2c2/google_cloud_storage-1.42.0-py2.py3-none-any.whl\n",
      "Requirement already satisfied: google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.34.0)\n",
      "Requirement already satisfied: google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.31.1)\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google.cloud.storage) (2.22.0)\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.3.3)\n",
      "Requirement already satisfied: google-cloud-core<3.0dev,>=1.6.0; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google.cloud.storage) (1.7.2)\n",
      "Requirement already satisfied: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.7.2)\n",
      "Requirement already satisfied: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (4.2.2)\n",
      "Requirement already satisfied: six>=1.9.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (1.16.0)\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.2.8)\n",
      "Requirement already satisfied: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (41.0.1)\n",
      "Requirement already satisfied: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2021.1)\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (1.53.0)\n",
      "Requirement already satisfied: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (21.0)\n",
      "Requirement already satisfied: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (3.17.3)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (1.25.11)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2.8)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (3.0.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google.cloud.storage) (2021.5.30)\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.1.2)\n",
      "Requirement already satisfied: pyasn1>=0.1.3 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from rsa<5,>=3.1.4; python_version >= \"3.6\"->google-auth<3.0dev,>=1.25.0; python_version >= \"3.6\"->google.cloud.storage) (0.4.8)\n",
      "Requirement already satisfied: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-api-core<3.0dev,>=1.29.0; python_version >= \"3.6\"->google.cloud.storage) (2.4.7)\n",
      "Requirement already satisfied: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (1.12.3)\n",
      "Requirement already satisfied: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=1.3.0; python_version >= \"3.6\"->google.cloud.storage) (2.19)\n",
      "Installing collected packages: google.cloud.storage\n",
      "\u001b[31mERROR: Could not install packages due to an EnvironmentError: [Errno 13] Permission denied: '/opt/tljh/user/lib/python3.7/site-packages/google_cloud_storage-1.42.0-py3.9-nspkg.pth'\n",
      "Consider using the `--user` option or check the permissions.\n",
      "\u001b[0m\n",
      "Collecting google-cloud-bigquery\n",
      "\u001b[33m  Cache entry deserialization failed, entry ignored\u001b[0m\n",
      "  Using cached https://files.pythonhosted.org/packages/fd/2a/d53b342d20e4b95ade480fa04977969166bb189bf0e909501e4fff86cb3f/google_cloud_bigquery-2.25.1-py2.py3-none-any.whl\n",
      "Collecting protobuf>=3.12.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/53/4e/e2db88d0bb0bda6a879eea62fddbaf813719ce3770d458bc5580512d9c95/protobuf-3.17.3-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.whl\n",
      "Collecting requests<3.0.0dev,>=2.18.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/92/96/144f70b972a9c0eabbd4391ef93ccd49d0f2747f4f6a2a2738e99e5adc65/requests-2.26.0-py2.py3-none-any.whl\n",
      "Collecting grpcio<2.0dev,>=1.38.1 (from google-cloud-bigquery)\n",
      "Collecting proto-plus>=1.10.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8c/72/6f3f4cdc5bb0294f8d7f3f8aacb617b4c3cb17554ed78f7e28009162c795/proto_plus-1.19.0-py3-none-any.whl\n",
      "Collecting google-cloud-core<3.0.0dev,>=1.4.1 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/42/b8/4ad3e423105d26a0c5eea54320aed57b7be060d1cd9faea53da713b408c6/google_cloud_core-2.0.0-py2.py3-none-any.whl\n",
      "Collecting google-resumable-media<3.0dev,>=0.6.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/2b/86/ff9b23ac7446ee224d0d7e5e64efc433aeef52b76662863873e03259ed0f/google_resumable_media-2.0.0-py2.py3-none-any.whl\n",
      "Collecting packaging>=14.3 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/3c/77/e2362b676dc5008d81be423070dd9577fa03be5da2ba1105811900fda546/packaging-21.0-py3-none-any.whl\n",
      "Collecting google-api-core[grpc]<3.0.0dev,>=1.29.0 (from google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8f/3c/6c64ce599505b5007e3b91ce9947030397f2bff591f52dbf708a0ccca4e1/google_api_core-2.0.1-py2.py3-none-any.whl\n",
      "Collecting six>=1.9 (from protobuf>=3.12.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d9/5a/e7c31adbe875f2abbb91bd84cf2dc52d792b5a01506781dbcf25c91daf11/six-1.16.0-py2.py3-none-any.whl\n",
      "Collecting urllib3<1.27,>=1.21.1 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/5f/64/43575537846896abac0b15c3e5ac678d787a4021e906703f1766bfb8ea11/urllib3-1.26.6-py2.py3-none-any.whl\n",
      "Collecting certifi>=2017.4.17 (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/05/1b/0a0dece0e8aa492a6ec9e4ad2fe366b511558cdc73fd3abc82ba7348e875/certifi-2021.5.30-py2.py3-none-any.whl\n",
      "Collecting charset-normalizer~=2.0.0; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/33/53/b7f6126a2b9fd878b025fe3c40266cfaad696f312165008ce045bffa3fe7/charset_normalizer-2.0.4-py3-none-any.whl\n",
      "Collecting idna<4,>=2.5; python_version >= \"3\" (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/d7/77/ff688d1504cdc4db2a938e2b7b9adee5dd52e34efbd2431051efc9984de9/idna-3.2-py3-none-any.whl\n",
      "Collecting google-auth<3.0dev,>=1.24.0 (from google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/48/27/4e5c7edb04bd5bb19fd1ab6add5a429f54bc45676ba93ccfd4e5331ba928/google_auth-2.0.2-py2.py3-none-any.whl\n",
      "Collecting google-crc32c<2.0dev,>=1.0 (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bb/36/dc98ccaffb0aae168100afe6954a5ce6f6c4ad11a31df507ccfe8b30b87a/google-crc32c-1.1.3.tar.gz\n",
      "Collecting pyparsing>=2.0.2 (from packaging>=14.3->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/8a/bb/488841f56197b13700afd5658fc279a2025a39e22449b7cf29864669b15d/pyparsing-2.4.7-py2.py3-none-any.whl\n",
      "Collecting googleapis-common-protos<2.0dev,>=1.6.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/55/08/796a6bc0b550e2b7116041c953d3d5100016abea106131d71e5651826e7b/googleapis_common_protos-1.53.0-py2.py3-none-any.whl\n",
      "Collecting setuptools>=40.3.0 (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bd/25/5bdf7f1adeebd4e3fa76b2e2f045ae53ee208e40a4231ad0f0c3007e4353/setuptools-57.4.0-py3-none-any.whl\n",
      "Collecting cachetools<5.0,>=2.0.0 (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/bf/28/c4f5796c67ad06bb91d98d543a5e01805c1ff065e08871f78e52d2a331ad/cachetools-4.2.2-py3-none-any.whl\n",
      "Collecting pyasn1-modules>=0.2.1 (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/95/de/214830a981892a3e286c3794f41ae67a4495df1108c3da8a9f62159b9a9d/pyasn1_modules-0.2.8-py2.py3-none-any.whl\n",
      "Collecting rsa<5,>=3.1.4 (from google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/e9/93/0c0f002031f18b53af7a6166103c02b9c0667be528944137cc954ec921b3/rsa-4.7.2-py3-none-any.whl\n",
      "Collecting pyasn1<0.5.0,>=0.4.6 (from pyasn1-modules>=0.2.1->google-auth<3.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery)\n",
      "  Using cached https://files.pythonhosted.org/packages/62/1e/a94a8d635fa3ce4cfc7f506003548d0a2447ae76fd5ca53932970fe3053f/pyasn1-0.4.8-py2.py3-none-any.whl\n",
      "Building wheels for collected packages: google-crc32c\n",
      "  Running setup.py bdist_wheel for google-crc32c ... \u001b[?25lerror\n",
      "  Complete output from command /usr/bin/python3 -u -c \"import setuptools, tokenize;__file__='/tmp/pip-build-v7pjmtp2/google-crc32c/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\\r\\n', '\\n');f.close();exec(compile(code, __file__, 'exec'))\" bdist_wheel -d /tmp/tmp34e2rpyypip-wheel- --python-tag cp36:\n",
      "  running bdist_wheel\n",
      "  running build\n",
      "  running build_py\n",
      "  creating build\n",
      "  creating build/lib.linux-x86_64-3.6\n",
      "  creating build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  copying src/google_crc32c/cext.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  copying src/google_crc32c/python.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  copying src/google_crc32c/__config__.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  copying src/google_crc32c/_checksum.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  copying src/google_crc32c/__init__.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "  running build_ext\n",
      "  building 'google_crc32c._crc32c' extension\n",
      "  creating build/temp.linux-x86_64-3.6\n",
      "  creating build/temp.linux-x86_64-3.6/src\n",
      "  creating build/temp.linux-x86_64-3.6/src/google_crc32c\n",
      "  x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -c src/google_crc32c/_crc32c.c -o build/temp.linux-x86_64-3.6/src/google_crc32c/_crc32c.o\n",
      "  src/google_crc32c/_crc32c.c:3:10: fatal error: crc32c/crc32c.h: No such file or directory\n",
      "   #include <crc32c/crc32c.h>\n",
      "            ^~~~~~~~~~~~~~~~~\n",
      "  compilation terminated.\n",
      "  error: command 'x86_64-linux-gnu-gcc' failed with exit status 1\n",
      "  \n",
      "  ----------------------------------------\n",
      "\u001b[31m  Failed building wheel for google-crc32c\u001b[0m\n",
      "\u001b[?25h  Running setup.py clean for google-crc32c\n",
      "Failed to build google-crc32c\n",
      "Installing collected packages: six, protobuf, urllib3, certifi, charset-normalizer, idna, requests, grpcio, proto-plus, cachetools, pyasn1, pyasn1-modules, rsa, setuptools, google-auth, googleapis-common-protos, google-api-core, google-cloud-core, google-crc32c, google-resumable-media, pyparsing, packaging, google-cloud-bigquery\n",
      "  Running setup.py install for google-crc32c ... \u001b[?25lerror\n",
      "    Complete output from command /usr/bin/python3 -u -c \"import setuptools, tokenize;__file__='/tmp/pip-build-v7pjmtp2/google-crc32c/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\\r\\n', '\\n');f.close();exec(compile(code, __file__, 'exec'))\" install --record /tmp/pip-mk9ba6p9-record/install-record.txt --single-version-externally-managed --compile --user --prefix=:\n",
      "    running install\n",
      "    running build\n",
      "    running build_py\n",
      "    creating build\n",
      "    creating build/lib.linux-x86_64-3.6\n",
      "    creating build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    copying src/google_crc32c/cext.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    copying src/google_crc32c/python.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    copying src/google_crc32c/__config__.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    copying src/google_crc32c/_checksum.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    copying src/google_crc32c/__init__.py -> build/lib.linux-x86_64-3.6/google_crc32c\n",
      "    running build_ext\n",
      "    building 'google_crc32c._crc32c' extension\n",
      "    creating build/temp.linux-x86_64-3.6\n",
      "    creating build/temp.linux-x86_64-3.6/src\n",
      "    creating build/temp.linux-x86_64-3.6/src/google_crc32c\n",
      "    x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/python3.6m -c src/google_crc32c/_crc32c.c -o build/temp.linux-x86_64-3.6/src/google_crc32c/_crc32c.o\n",
      "    src/google_crc32c/_crc32c.c:3:10: fatal error: crc32c/crc32c.h: No such file or directory\n",
      "     #include <crc32c/crc32c.h>\n",
      "              ^~~~~~~~~~~~~~~~~\n",
      "    compilation terminated.\n",
      "    error: command 'x86_64-linux-gnu-gcc' failed with exit status 1\n",
      "    \n",
      "    ----------------------------------------\n",
      "\u001b[31mCommand \"/usr/bin/python3 -u -c \"import setuptools, tokenize;__file__='/tmp/pip-build-v7pjmtp2/google-crc32c/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\\r\\n', '\\n');f.close();exec(compile(code, __file__, 'exec'))\" install --record /tmp/pip-mk9ba6p9-record/install-record.txt --single-version-externally-managed --compile --user --prefix=\" failed with error code 1 in /tmp/pip-build-v7pjmtp2/google-crc32c/\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: google in /home/jupyter-astine/.local/lib/python3.7/site-packages (3.0.0)\n",
      "Requirement already satisfied: beautifulsoup4 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google) (4.9.3)\n",
      "Requirement already satisfied: soupsieve>1.2; python_version >= \"3.0\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from beautifulsoup4->google) (2.2.1)\n",
      "Requirement already up-to-date: google-cloud-bigquery[bqstorage,pandas] in /home/jupyter-astine/.local/lib/python3.7/site-packages (2.25.1)\n",
      "Requirement already satisfied, skipping upgrade: google-resumable-media<3.0dev,>=0.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.3)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-core<3.0.0dev,>=1.4.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.7.2)\n",
      "Requirement already satisfied, skipping upgrade: google-api-core[grpc]<3.0.0dev,>=1.29.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.31.1)\n",
      "Requirement already satisfied, skipping upgrade: packaging>=14.3 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (21.0)\n",
      "Requirement already satisfied, skipping upgrade: grpcio<2.0dev,>=1.38.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.39.0)\n",
      "Requirement already satisfied, skipping upgrade: requests<3.0.0dev,>=2.18.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.22.0)\n",
      "Requirement already satisfied, skipping upgrade: proto-plus>=1.10.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.19.0)\n",
      "Requirement already satisfied, skipping upgrade: protobuf>=3.12.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (3.17.3)\n",
      "Requirement already satisfied, skipping upgrade: google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (2.6.3)\n",
      "Requirement already satisfied, skipping upgrade: pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (5.0.0)\n",
      "Requirement already satisfied, skipping upgrade: pandas>=0.23.0; extra == \"pandas\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery[bqstorage,pandas]) (1.3.1)\n",
      "Requirement already satisfied, skipping upgrade: six>=1.4.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.16.0)\n",
      "Requirement already satisfied, skipping upgrade: google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.1.2)\n",
      "Requirement already satisfied, skipping upgrade: google-auth<2.0dev,>=1.24.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (1.34.0)\n",
      "Requirement already satisfied, skipping upgrade: googleapis-common-protos<2.0dev,>=1.6.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (1.53.0)\n",
      "Requirement already satisfied, skipping upgrade: pytz in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (2021.1)\n",
      "Requirement already satisfied, skipping upgrade: setuptools>=40.3.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-api-core[grpc]<3.0.0dev,>=1.29.0->google-cloud-bigquery[bqstorage,pandas]) (41.0.1)\n",
      "Requirement already satisfied, skipping upgrade: pyparsing>=2.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from packaging>=14.3->google-cloud-bigquery[bqstorage,pandas]) (2.4.7)\n",
      "Requirement already satisfied, skipping upgrade: certifi>=2017.4.17 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2021.5.30)\n",
      "Requirement already satisfied, skipping upgrade: chardet<3.1.0,>=3.0.2 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (3.0.4)\n",
      "Requirement already satisfied, skipping upgrade: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (1.25.11)\n",
      "Requirement already satisfied, skipping upgrade: idna<2.9,>=2.5 in /opt/tljh/user/lib/python3.7/site-packages (from requests<3.0.0dev,>=2.18.0->google-cloud-bigquery[bqstorage,pandas]) (2.8)\n",
      "Requirement already satisfied, skipping upgrade: libcst>=0.2.5 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.3.20)\n",
      "Requirement already satisfied, skipping upgrade: numpy>=1.16.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyarrow<6.0dev,>=3.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (1.21.1)\n",
      "Requirement already satisfied, skipping upgrade: python-dateutil>=2.7.3 in /opt/tljh/user/lib/python3.7/site-packages (from pandas>=0.23.0; extra == \"pandas\"->google-cloud-bigquery[bqstorage,pandas]) (2.8.2)\n",
      "Requirement already satisfied, skipping upgrade: cffi>=1.0.0 in /opt/tljh/user/lib/python3.7/site-packages (from google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (1.12.3)\n",
      "Requirement already satisfied, skipping upgrade: cachetools<5.0,>=2.0.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.2.2)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1-modules>=0.2.1 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.2.8)\n",
      "Requirement already satisfied, skipping upgrade: rsa<5,>=3.1.4; python_version >= \"3.6\" in /home/jupyter-astine/.local/lib/python3.7/site-packages (from google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (4.7.2)\n",
      "Requirement already satisfied, skipping upgrade: pyyaml>=5.2 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (5.4.1)\n",
      "Requirement already satisfied, skipping upgrade: typing-extensions>=3.7.4.2 in /opt/tljh/user/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (3.10.0.0)\n",
      "Requirement already satisfied, skipping upgrade: typing-inspect>=0.4.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.7.1)\n",
      "Requirement already satisfied, skipping upgrade: pycparser in /opt/tljh/user/lib/python3.7/site-packages (from cffi>=1.0.0->google-crc32c<2.0dev,>=1.0; python_version >= \"3.5\"->google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery[bqstorage,pandas]) (2.19)\n",
      "Requirement already satisfied, skipping upgrade: pyasn1<0.5.0,>=0.4.6 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from pyasn1-modules>=0.2.1->google-auth<2.0dev,>=1.24.0->google-cloud-core<3.0.0dev,>=1.4.1->google-cloud-bigquery[bqstorage,pandas]) (0.4.8)\n",
      "Requirement already satisfied, skipping upgrade: mypy-extensions>=0.3.0 in /home/jupyter-astine/.local/lib/python3.7/site-packages (from typing-inspect>=0.4.0->libcst>=0.2.5->google-cloud-bigquery-storage<3.0.0dev,>=2.0.0; extra == \"bqstorage\"->google-cloud-bigquery[bqstorage,pandas]) (0.4.3)\n"
     ]
    }
   ],
   "source": [
    "!pip install google.cloud.bigquery\n",
    "!pip install google.cloud.storage\n",
    "!pip3 install --upgrade google-cloud-bigquery\n",
    "\n",
    "!pip install google --user\n",
    "!pip install --upgrade 'google-cloud-bigquery[bqstorage,pandas]' --user\n",
    "\n",
    "%load_ext google.cloud.bigquery"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b0c3c81",
   "metadata": {},
   "source": [
    "## Basic Query\n",
    "Here is a simple query to find 5 runs submitted for the organsim 'Homo sapiens' from the nih-sra-datastore project. This query searches for all columns in the metadata table from the sra database in the project. We will filter for only the rows where the organism name is 'Homo sapiens' and limit the output to only the first 5 rows by using the LIMIT function. One limitation of our view in Jupyter is only 20 columns are displayed. At the bottom of the results you will see 5 rows X 36 columns. But if you look closely there are 16 columns that are not visible in the middle of the output chart."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "81be3e67",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 2/2 [00:00<00:00, 1737.49query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  7.33rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>geo_loc_name_country_continent_calc</th>\n",
       "      <th>geo_loc_name_sam</th>\n",
       "      <th>ena_first_public_run</th>\n",
       "      <th>ena_last_update_run</th>\n",
       "      <th>sample_name_sam</th>\n",
       "      <th>datastore_filetype</th>\n",
       "      <th>datastore_provider</th>\n",
       "      <th>datastore_region</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR15278393</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>EMORY UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11583084</td>\n",
       "      <td>2815176</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '9518138700'}, {'k': 'byt...</td>\n",
       "      <td>{\"bases\": 9518138700, \"bytes\": 3239993193, \"ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR15278347</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>EMORY UNIVERSITY</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11583130</td>\n",
       "      <td>919115</td>\n",
       "      <td>Illumina HiSeq 1000</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '9340662300'}, {'k': 'byt...</td>\n",
       "      <td>{\"bases\": 9340662300, \"bytes\": 2922928789, \"ag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR15065199</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375204</td>\n",
       "      <td>GSM5412973</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '188792856'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 188792856, \"bytes\": 85176148, \"age_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR15065057</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375346</td>\n",
       "      <td>GSM5412928</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '409589460'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 409589460, \"bytes\": 186427640, \"age_...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR15065033</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>THE OPEN UNIVERSITY OF ISRAEL</td>\n",
       "      <td>public</td>\n",
       "      <td>SRX11375370</td>\n",
       "      <td>GSM5412909</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[fastq, sra]</td>\n",
       "      <td>[gs, ncbi, s3]</td>\n",
       "      <td>[gs.US, ncbi.public, s3.us-east-1]</td>\n",
       "      <td>[{'k': 'bases', 'v': '245428848'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 245428848, \"bytes\": 103593882, \"age_...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc assay_type                    center_name consent   experiment  \\\n",
       "0  SRR15278393    RNA-Seq               EMORY UNIVERSITY  public  SRX11583084   \n",
       "1  SRR15278347    RNA-Seq               EMORY UNIVERSITY  public  SRX11583130   \n",
       "2  SRR15065199   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375204   \n",
       "3  SRR15065057   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375346   \n",
       "4  SRR15065033   AMPLICON  THE OPEN UNIVERSITY OF ISRAEL  public  SRX11375370   \n",
       "\n",
       "  sample_name           instrument librarylayout libraryselection  \\\n",
       "0     2815176  Illumina HiSeq 1000        PAIRED            PolyA   \n",
       "1      919115  Illumina HiSeq 1000        PAIRED            PolyA   \n",
       "2  GSM5412973  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "3  GSM5412928  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "4  GSM5412909  Illumina HiSeq 2500        PAIRED              PCR   \n",
       "\n",
       "    librarysource  ... geo_loc_name_country_continent_calc geo_loc_name_sam  \\\n",
       "0  TRANSCRIPTOMIC  ...                                None               []   \n",
       "1  TRANSCRIPTOMIC  ...                                None               []   \n",
       "2     METAGENOMIC  ...                                None               []   \n",
       "3     METAGENOMIC  ...                                None               []   \n",
       "4     METAGENOMIC  ...                                None               []   \n",
       "\n",
       "  ena_first_public_run ena_last_update_run sample_name_sam datastore_filetype  \\\n",
       "0                   []                  []              []       [fastq, sra]   \n",
       "1                   []                  []              []       [fastq, sra]   \n",
       "2                   []                  []              []       [fastq, sra]   \n",
       "3                   []                  []              []       [fastq, sra]   \n",
       "4                   []                  []              []       [fastq, sra]   \n",
       "\n",
       "  datastore_provider                    datastore_region  \\\n",
       "0     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "1     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "2     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "3     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "4     [gs, ncbi, s3]  [gs.US, ncbi.public, s3.us-east-1]   \n",
       "\n",
       "                                          attributes  \\\n",
       "0  [{'k': 'bases', 'v': '9518138700'}, {'k': 'byt...   \n",
       "1  [{'k': 'bases', 'v': '9340662300'}, {'k': 'byt...   \n",
       "2  [{'k': 'bases', 'v': '188792856'}, {'k': 'byte...   \n",
       "3  [{'k': 'bases', 'v': '409589460'}, {'k': 'byte...   \n",
       "4  [{'k': 'bases', 'v': '245428848'}, {'k': 'byte...   \n",
       "\n",
       "                                               jattr  \n",
       "0  {\"bases\": 9518138700, \"bytes\": 3239993193, \"ag...  \n",
       "1  {\"bases\": 9340662300, \"bytes\": 2922928789, \"ag...  \n",
       "2  {\"bases\": 188792856, \"bytes\": 85176148, \"age_s...  \n",
       "3  {\"bases\": 409589460, \"bytes\": 186427640, \"age_...  \n",
       "4  {\"bases\": 245428848, \"bytes\": 103593882, \"age_...  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT *\n",
    "FROM `nih-sra-datastore.sra.metadata`\n",
    "WHERE organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fea8d95",
   "metadata": {},
   "source": [
    "## Searching Other Tables\n",
    "The second database in the nih-sra-datastore project contains data from the SRA Taxonomy Analysis Tool (STAT) as well as some from the Taxonomy database. The analysis will use the Taxonomy Database Identifier (tax ID) in the results. You can find the tax_id for any entry in the taxonomy table using a search like the one below. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "21842a8a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 651.90query/s]                          \n",
      "Downloading: 100%|██████████| 1/1 [00:00<00:00,  1.51rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>names</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>ileft</th>\n",
       "      <th>iright</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   tax_id  parent_id     rank      sci_name  \\\n",
       "0    9606       9605  species  Homo sapiens   \n",
       "\n",
       "                                               names  ilevel   ileft  iright  \n",
       "0  [{'name': 'Home sapiens', 'name_class': 'missp...      30  212230  212231  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT * \n",
    "FROM `nih-sra-datastore.sra_tax_analysis_tool.taxonomy`\n",
    "WHERE sci_name = 'Homo sapiens'\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebc8441a",
   "metadata": {},
   "source": [
    "## Finding How the Database is Organized and Listing Columns\n",
    "We can use the query below to get a listing of the columns in the database schema. This can be useful to see all the names of columns in a table as well as seeing what type of data is in them. The SQL format requires single quotes around string values in searches but no quotes for integers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f1207cb2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 1/1 [00:00<00:00, 971.35query/s]                          \n",
      "Downloading: 100%|██████████| 36/36 [00:00<00:00, 44.26rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>table_catalog</th>\n",
       "      <th>table_schema</th>\n",
       "      <th>table_name</th>\n",
       "      <th>column_name</th>\n",
       "      <th>ordinal_position</th>\n",
       "      <th>is_nullable</th>\n",
       "      <th>data_type</th>\n",
       "      <th>is_generated</th>\n",
       "      <th>generation_expression</th>\n",
       "      <th>is_stored</th>\n",
       "      <th>is_hidden</th>\n",
       "      <th>is_updatable</th>\n",
       "      <th>is_system_defined</th>\n",
       "      <th>is_partitioning_column</th>\n",
       "      <th>clustering_ordinal_position</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>acc</td>\n",
       "      <td>1</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>2</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>center_name</td>\n",
       "      <td>3</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>consent</td>\n",
       "      <td>4</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>experiment</td>\n",
       "      <td>5</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>sample_name</td>\n",
       "      <td>6</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>instrument</td>\n",
       "      <td>7</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>librarylayout</td>\n",
       "      <td>8</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>libraryselection</td>\n",
       "      <td>9</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>librarysource</td>\n",
       "      <td>10</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>platform</td>\n",
       "      <td>11</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>sample_acc</td>\n",
       "      <td>12</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>biosample</td>\n",
       "      <td>13</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>organism</td>\n",
       "      <td>14</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>sra_study</td>\n",
       "      <td>15</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>releasedate</td>\n",
       "      <td>16</td>\n",
       "      <td>YES</td>\n",
       "      <td>TIMESTAMP</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>bioproject</td>\n",
       "      <td>17</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>mbytes</td>\n",
       "      <td>18</td>\n",
       "      <td>YES</td>\n",
       "      <td>INT64</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>loaddate</td>\n",
       "      <td>19</td>\n",
       "      <td>YES</td>\n",
       "      <td>TIMESTAMP</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>avgspotlen</td>\n",
       "      <td>20</td>\n",
       "      <td>YES</td>\n",
       "      <td>INT64</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>mbases</td>\n",
       "      <td>21</td>\n",
       "      <td>YES</td>\n",
       "      <td>INT64</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>insertsize</td>\n",
       "      <td>22</td>\n",
       "      <td>YES</td>\n",
       "      <td>INT64</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>library_name</td>\n",
       "      <td>23</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>biosamplemodel_sam</td>\n",
       "      <td>24</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>collection_date_sam</td>\n",
       "      <td>25</td>\n",
       "      <td>YES</td>\n",
       "      <td>DATE</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>geo_loc_name_country_calc</td>\n",
       "      <td>26</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>geo_loc_name_country_continent_calc</td>\n",
       "      <td>27</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>geo_loc_name_sam</td>\n",
       "      <td>28</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>ena_first_public_run</td>\n",
       "      <td>29</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>ena_last_update_run</td>\n",
       "      <td>30</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>sample_name_sam</td>\n",
       "      <td>31</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>datastore_filetype</td>\n",
       "      <td>32</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>datastore_provider</td>\n",
       "      <td>33</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>datastore_region</td>\n",
       "      <td>34</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRING&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>attributes</td>\n",
       "      <td>35</td>\n",
       "      <td>NO</td>\n",
       "      <td>ARRAY&lt;STRUCT&lt;k STRING, v STRING&gt;&gt;</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>nih-sra-datastore</td>\n",
       "      <td>sra</td>\n",
       "      <td>metadata</td>\n",
       "      <td>jattr</td>\n",
       "      <td>36</td>\n",
       "      <td>YES</td>\n",
       "      <td>STRING</td>\n",
       "      <td>NEVER</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>None</td>\n",
       "      <td>NO</td>\n",
       "      <td>NO</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        table_catalog table_schema table_name  \\\n",
       "0   nih-sra-datastore          sra   metadata   \n",
       "1   nih-sra-datastore          sra   metadata   \n",
       "2   nih-sra-datastore          sra   metadata   \n",
       "3   nih-sra-datastore          sra   metadata   \n",
       "4   nih-sra-datastore          sra   metadata   \n",
       "5   nih-sra-datastore          sra   metadata   \n",
       "6   nih-sra-datastore          sra   metadata   \n",
       "7   nih-sra-datastore          sra   metadata   \n",
       "8   nih-sra-datastore          sra   metadata   \n",
       "9   nih-sra-datastore          sra   metadata   \n",
       "10  nih-sra-datastore          sra   metadata   \n",
       "11  nih-sra-datastore          sra   metadata   \n",
       "12  nih-sra-datastore          sra   metadata   \n",
       "13  nih-sra-datastore          sra   metadata   \n",
       "14  nih-sra-datastore          sra   metadata   \n",
       "15  nih-sra-datastore          sra   metadata   \n",
       "16  nih-sra-datastore          sra   metadata   \n",
       "17  nih-sra-datastore          sra   metadata   \n",
       "18  nih-sra-datastore          sra   metadata   \n",
       "19  nih-sra-datastore          sra   metadata   \n",
       "20  nih-sra-datastore          sra   metadata   \n",
       "21  nih-sra-datastore          sra   metadata   \n",
       "22  nih-sra-datastore          sra   metadata   \n",
       "23  nih-sra-datastore          sra   metadata   \n",
       "24  nih-sra-datastore          sra   metadata   \n",
       "25  nih-sra-datastore          sra   metadata   \n",
       "26  nih-sra-datastore          sra   metadata   \n",
       "27  nih-sra-datastore          sra   metadata   \n",
       "28  nih-sra-datastore          sra   metadata   \n",
       "29  nih-sra-datastore          sra   metadata   \n",
       "30  nih-sra-datastore          sra   metadata   \n",
       "31  nih-sra-datastore          sra   metadata   \n",
       "32  nih-sra-datastore          sra   metadata   \n",
       "33  nih-sra-datastore          sra   metadata   \n",
       "34  nih-sra-datastore          sra   metadata   \n",
       "35  nih-sra-datastore          sra   metadata   \n",
       "\n",
       "                            column_name  ordinal_position is_nullable  \\\n",
       "0                                   acc                 1         YES   \n",
       "1                            assay_type                 2         YES   \n",
       "2                           center_name                 3         YES   \n",
       "3                               consent                 4         YES   \n",
       "4                            experiment                 5         YES   \n",
       "5                           sample_name                 6         YES   \n",
       "6                            instrument                 7         YES   \n",
       "7                         librarylayout                 8         YES   \n",
       "8                      libraryselection                 9         YES   \n",
       "9                         librarysource                10         YES   \n",
       "10                             platform                11         YES   \n",
       "11                           sample_acc                12         YES   \n",
       "12                            biosample                13         YES   \n",
       "13                             organism                14         YES   \n",
       "14                            sra_study                15         YES   \n",
       "15                          releasedate                16         YES   \n",
       "16                           bioproject                17         YES   \n",
       "17                               mbytes                18         YES   \n",
       "18                             loaddate                19         YES   \n",
       "19                           avgspotlen                20         YES   \n",
       "20                               mbases                21         YES   \n",
       "21                           insertsize                22         YES   \n",
       "22                         library_name                23         YES   \n",
       "23                   biosamplemodel_sam                24          NO   \n",
       "24                  collection_date_sam                25         YES   \n",
       "25            geo_loc_name_country_calc                26         YES   \n",
       "26  geo_loc_name_country_continent_calc                27         YES   \n",
       "27                     geo_loc_name_sam                28          NO   \n",
       "28                 ena_first_public_run                29          NO   \n",
       "29                  ena_last_update_run                30          NO   \n",
       "30                      sample_name_sam                31          NO   \n",
       "31                   datastore_filetype                32          NO   \n",
       "32                   datastore_provider                33          NO   \n",
       "33                     datastore_region                34          NO   \n",
       "34                           attributes                35          NO   \n",
       "35                                jattr                36         YES   \n",
       "\n",
       "                            data_type is_generated generation_expression  \\\n",
       "0                              STRING        NEVER                  None   \n",
       "1                              STRING        NEVER                  None   \n",
       "2                              STRING        NEVER                  None   \n",
       "3                              STRING        NEVER                  None   \n",
       "4                              STRING        NEVER                  None   \n",
       "5                              STRING        NEVER                  None   \n",
       "6                              STRING        NEVER                  None   \n",
       "7                              STRING        NEVER                  None   \n",
       "8                              STRING        NEVER                  None   \n",
       "9                              STRING        NEVER                  None   \n",
       "10                             STRING        NEVER                  None   \n",
       "11                             STRING        NEVER                  None   \n",
       "12                             STRING        NEVER                  None   \n",
       "13                             STRING        NEVER                  None   \n",
       "14                             STRING        NEVER                  None   \n",
       "15                          TIMESTAMP        NEVER                  None   \n",
       "16                             STRING        NEVER                  None   \n",
       "17                              INT64        NEVER                  None   \n",
       "18                          TIMESTAMP        NEVER                  None   \n",
       "19                              INT64        NEVER                  None   \n",
       "20                              INT64        NEVER                  None   \n",
       "21                              INT64        NEVER                  None   \n",
       "22                             STRING        NEVER                  None   \n",
       "23                      ARRAY<STRING>        NEVER                  None   \n",
       "24                               DATE        NEVER                  None   \n",
       "25                             STRING        NEVER                  None   \n",
       "26                             STRING        NEVER                  None   \n",
       "27                      ARRAY<STRING>        NEVER                  None   \n",
       "28                      ARRAY<STRING>        NEVER                  None   \n",
       "29                      ARRAY<STRING>        NEVER                  None   \n",
       "30                      ARRAY<STRING>        NEVER                  None   \n",
       "31                      ARRAY<STRING>        NEVER                  None   \n",
       "32                      ARRAY<STRING>        NEVER                  None   \n",
       "33                      ARRAY<STRING>        NEVER                  None   \n",
       "34  ARRAY<STRUCT<k STRING, v STRING>>        NEVER                  None   \n",
       "35                             STRING        NEVER                  None   \n",
       "\n",
       "   is_stored is_hidden is_updatable is_system_defined is_partitioning_column  \\\n",
       "0       None        NO         None                NO                     NO   \n",
       "1       None        NO         None                NO                     NO   \n",
       "2       None        NO         None                NO                     NO   \n",
       "3       None        NO         None                NO                     NO   \n",
       "4       None        NO         None                NO                     NO   \n",
       "5       None        NO         None                NO                     NO   \n",
       "6       None        NO         None                NO                     NO   \n",
       "7       None        NO         None                NO                     NO   \n",
       "8       None        NO         None                NO                     NO   \n",
       "9       None        NO         None                NO                     NO   \n",
       "10      None        NO         None                NO                     NO   \n",
       "11      None        NO         None                NO                     NO   \n",
       "12      None        NO         None                NO                     NO   \n",
       "13      None        NO         None                NO                     NO   \n",
       "14      None        NO         None                NO                     NO   \n",
       "15      None        NO         None                NO                     NO   \n",
       "16      None        NO         None                NO                     NO   \n",
       "17      None        NO         None                NO                     NO   \n",
       "18      None        NO         None                NO                     NO   \n",
       "19      None        NO         None                NO                     NO   \n",
       "20      None        NO         None                NO                     NO   \n",
       "21      None        NO         None                NO                     NO   \n",
       "22      None        NO         None                NO                     NO   \n",
       "23      None        NO         None                NO                     NO   \n",
       "24      None        NO         None                NO                     NO   \n",
       "25      None        NO         None                NO                     NO   \n",
       "26      None        NO         None                NO                     NO   \n",
       "27      None        NO         None                NO                     NO   \n",
       "28      None        NO         None                NO                     NO   \n",
       "29      None        NO         None                NO                     NO   \n",
       "30      None        NO         None                NO                     NO   \n",
       "31      None        NO         None                NO                     NO   \n",
       "32      None        NO         None                NO                     NO   \n",
       "33      None        NO         None                NO                     NO   \n",
       "34      None        NO         None                NO                     NO   \n",
       "35      None        NO         None                NO                     NO   \n",
       "\n",
       "    clustering_ordinal_position  \n",
       "0                           NaN  \n",
       "1                           NaN  \n",
       "2                           NaN  \n",
       "3                           NaN  \n",
       "4                           NaN  \n",
       "5                           NaN  \n",
       "6                           NaN  \n",
       "7                           NaN  \n",
       "8                           NaN  \n",
       "9                           NaN  \n",
       "10                          NaN  \n",
       "11                          NaN  \n",
       "12                          NaN  \n",
       "13                          NaN  \n",
       "14                          NaN  \n",
       "15                          NaN  \n",
       "16                          NaN  \n",
       "17                          NaN  \n",
       "18                          NaN  \n",
       "19                          NaN  \n",
       "20                          NaN  \n",
       "21                          NaN  \n",
       "22                          NaN  \n",
       "23                          NaN  \n",
       "24                          NaN  \n",
       "25                          NaN  \n",
       "26                          NaN  \n",
       "27                          NaN  \n",
       "28                          NaN  \n",
       "29                          NaN  \n",
       "30                          NaN  \n",
       "31                          NaN  \n",
       "32                          NaN  \n",
       "33                          NaN  \n",
       "34                          NaN  \n",
       "35                          NaN  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT * FROM `nih-sra-datastore.sra`.INFORMATION_SCHEMA.COLUMNS\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33f7938d",
   "metadata": {},
   "source": [
    "## Joining Two (or more) Tables\n",
    "Using a JOIN command will allow you to combine two different tables to find the information you are looking for. This allows data to be stored most efficiently in the database but also allows a user to view the information in a way they find easiest to understand. Below we'll combine the two queries from before. When specifying multiple tables to search we can also give an abbreviated name to the table ('meta' and 'tax' in the query below) to make the rest of the query easier to type and read.\n",
    "When joining two or more tables we will need to tell the database which columns in the two tables are expected to align or have the same content. We need a field that can be used to link each table. This value is what we are joining \"on\" in the statement. In this case we are using 'organism' from the metadata table and 'sci_name' from the taxonomy table. Often in a database the columns used to join two tables will have the same name. But that is not always the case, this is one example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "45a293a7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 3118.44query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  6.81rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>center_name</th>\n",
       "      <th>consent</th>\n",
       "      <th>experiment</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>instrument</th>\n",
       "      <th>librarylayout</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>...</th>\n",
       "      <th>attributes</th>\n",
       "      <th>jattr</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>parent_id</th>\n",
       "      <th>rank</th>\n",
       "      <th>sci_name</th>\n",
       "      <th>names</th>\n",
       "      <th>ilevel</th>\n",
       "      <th>ileft</th>\n",
       "      <th>iright</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR2511560</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530975</td>\n",
       "      <td>SAMEA1116178</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '171494517'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 171494517, \"bytes\": 140015172, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR2511826</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2531241</td>\n",
       "      <td>SAMEA1116444</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '163581768'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 163581768, \"bytes\": 132336304, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR2511209</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530624</td>\n",
       "      <td>SAMEA1115827</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '191747089'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 191747089, \"bytes\": 160929514, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR2511604</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2531019</td>\n",
       "      <td>SAMEA1116222</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '206745187'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 206745187, \"bytes\": 165491895, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR2511369</td>\n",
       "      <td>AMPLICON</td>\n",
       "      <td>BIOINFORMATICS LAB, INNOVATION CENTER</td>\n",
       "      <td>public</td>\n",
       "      <td>ERX2530784</td>\n",
       "      <td>SAMEA1115987</td>\n",
       "      <td>Ion S5</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>PCR</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>...</td>\n",
       "      <td>[{'k': 'bases', 'v': '167396357'}, {'k': 'byte...</td>\n",
       "      <td>{\"bases\": 167396357, \"bytes\": 134853975, \"alia...</td>\n",
       "      <td>9606</td>\n",
       "      <td>9605</td>\n",
       "      <td>species</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>[{'name': 'Home sapiens', 'name_class': 'missp...</td>\n",
       "      <td>30</td>\n",
       "      <td>212230</td>\n",
       "      <td>212231</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 44 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          acc assay_type                            center_name consent  \\\n",
       "0  ERR2511560   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "1  ERR2511826   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "2  ERR2511209   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "3  ERR2511604   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "4  ERR2511369   AMPLICON  BIOINFORMATICS LAB, INNOVATION CENTER  public   \n",
       "\n",
       "   experiment   sample_name instrument librarylayout libraryselection  \\\n",
       "0  ERX2530975  SAMEA1116178     Ion S5        SINGLE              PCR   \n",
       "1  ERX2531241  SAMEA1116444     Ion S5        SINGLE              PCR   \n",
       "2  ERX2530624  SAMEA1115827     Ion S5        SINGLE              PCR   \n",
       "3  ERX2531019  SAMEA1116222     Ion S5        SINGLE              PCR   \n",
       "4  ERX2530784  SAMEA1115987     Ion S5        SINGLE              PCR   \n",
       "\n",
       "  librarysource  ...                                         attributes  \\\n",
       "0   METAGENOMIC  ...  [{'k': 'bases', 'v': '171494517'}, {'k': 'byte...   \n",
       "1   METAGENOMIC  ...  [{'k': 'bases', 'v': '163581768'}, {'k': 'byte...   \n",
       "2   METAGENOMIC  ...  [{'k': 'bases', 'v': '191747089'}, {'k': 'byte...   \n",
       "3   METAGENOMIC  ...  [{'k': 'bases', 'v': '206745187'}, {'k': 'byte...   \n",
       "4   METAGENOMIC  ...  [{'k': 'bases', 'v': '167396357'}, {'k': 'byte...   \n",
       "\n",
       "                                               jattr tax_id parent_id  \\\n",
       "0  {\"bases\": 171494517, \"bytes\": 140015172, \"alia...   9606      9605   \n",
       "1  {\"bases\": 163581768, \"bytes\": 132336304, \"alia...   9606      9605   \n",
       "2  {\"bases\": 191747089, \"bytes\": 160929514, \"alia...   9606      9605   \n",
       "3  {\"bases\": 206745187, \"bytes\": 165491895, \"alia...   9606      9605   \n",
       "4  {\"bases\": 167396357, \"bytes\": 134853975, \"alia...   9606      9605   \n",
       "\n",
       "      rank      sci_name                                              names  \\\n",
       "0  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "1  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "2  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "3  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "4  species  Homo sapiens  [{'name': 'Home sapiens', 'name_class': 'missp...   \n",
       "\n",
       "   ilevel   ileft  iright  \n",
       "0      30  212230  212231  \n",
       "1      30  212230  212231  \n",
       "2      30  212230  212231  \n",
       "3      30  212230  212231  \n",
       "4      30  212230  212231  \n",
       "\n",
       "[5 rows x 44 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT *\n",
    "FROM `nih-sra-datastore.sra.metadata` meta \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.taxonomy` tax\n",
    "    ON meta.organism = tax.sci_name\n",
    "WHERE tax.tax_id = 9606\n",
    "    AND meta.organism = 'Homo sapiens'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54f6a913",
   "metadata": {},
   "source": [
    "## Selecting Fewer Columns\n",
    "Using * in the select statement will show all the columns in the table. Often we don't need or want all of the columns so we can show only the columns we are interested in. We do this by listing the columns we want in the select statement."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "b89a5690",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 3520.93query/s]                        \n",
      "Downloading: 100%|██████████| 5/5 [00:00<00:00,  6.10rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>assay_type</th>\n",
       "      <th>instrument</th>\n",
       "      <th>libraryselection</th>\n",
       "      <th>librarysource</th>\n",
       "      <th>sci_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR12245574</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>HiSeq X Ten</td>\n",
       "      <td>PolyA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR15667241</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR15667236</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR14683905</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR14683915</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>cDNA</td>\n",
       "      <td>TRANSCRIPTOMIC</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           acc assay_type             instrument libraryselection  \\\n",
       "0  SRR12245574    RNA-Seq            HiSeq X Ten            PolyA   \n",
       "1  SRR15667241    RNA-Seq    Illumina HiSeq 2500             cDNA   \n",
       "2  SRR15667236    RNA-Seq    Illumina HiSeq 2500             cDNA   \n",
       "3  SRR14683905    RNA-Seq  Illumina NovaSeq 6000             cDNA   \n",
       "4  SRR14683915    RNA-Seq  Illumina NovaSeq 6000             cDNA   \n",
       "\n",
       "    librarysource      sci_name  \n",
       "0  TRANSCRIPTOMIC  Homo sapiens  \n",
       "1  TRANSCRIPTOMIC  Homo sapiens  \n",
       "2  TRANSCRIPTOMIC  Homo sapiens  \n",
       "3  TRANSCRIPTOMIC  Homo sapiens  \n",
       "4  TRANSCRIPTOMIC  Homo sapiens  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT meta.acc, meta.assay_type, meta.instrument, meta.libraryselection, meta.librarysource, tax.sci_name\n",
    "FROM `nih-sra-datastore.sra.metadata` meta \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.taxonomy` tax\n",
    "    ON meta.organism = tax.sci_name\n",
    "WHERE tax.tax_id = 9606\n",
    "    AND meta.organism = 'Homo sapiens'\n",
    "    AND meta.assay_type = 'RNA-Seq'\n",
    "    AND meta.consent = 'public'\n",
    "LIMIT 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc2b0ad1",
   "metadata": {},
   "source": [
    "## Using Where Clauses to Find Data of Interest\n",
    "Now let's look at an example for someone who wants to do an alignment. We will combine some ideas from before as well as ad some math into the search. If you wanted to use the PGAP package from NCBI, you would want data that has a high percentage (we'll use 90% as our minimum) of only one species to generate an assembly and annotation for. We can use the output from STAT to do this. \n",
    "\n",
    "We want to make an assembly and annotation for Listeria monocytogenes. \n",
    "1. We'll use the taxid (1639) to search the tax_id column in the tax_analysis table. \n",
    "2. To know what proportion of the spots have been identified to the taxid we are searching for, we also need to use the total analyzed_spot_count from the tax_analysis_info table for each run. \n",
    "3. We will need to join these two tables on the run accession (acc) to run this query.\n",
    "\n",
    "Finally we can sort by the proportion of the spots that were identified as being from the taxid we want to search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "969f780b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 4/4 [00:00<00:00, 3391.39query/s]                        \n",
      "Downloading: 100%|██████████| 16153/16153 [00:01<00:00, 15628.64rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>acc</th>\n",
       "      <th>name</th>\n",
       "      <th>analyzed_spot_count</th>\n",
       "      <th>self_count</th>\n",
       "      <th>proportion</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR1817002</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>608593</td>\n",
       "      <td>594332</td>\n",
       "      <td>0.976567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR1816986</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>692180</td>\n",
       "      <td>673720</td>\n",
       "      <td>0.973331</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ERR1816981</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>630954</td>\n",
       "      <td>613122</td>\n",
       "      <td>0.971738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ERR1816996</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>593342</td>\n",
       "      <td>575904</td>\n",
       "      <td>0.970611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ERR1431120</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>3249346</td>\n",
       "      <td>3153217</td>\n",
       "      <td>0.970416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16148</th>\n",
       "      <td>SRR10665812</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>510003</td>\n",
       "      <td>459009</td>\n",
       "      <td>0.900012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16149</th>\n",
       "      <td>SRR2071980</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>1336913</td>\n",
       "      <td>1203238</td>\n",
       "      <td>0.900012</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16150</th>\n",
       "      <td>SRR6038661</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>749339</td>\n",
       "      <td>674410</td>\n",
       "      <td>0.900007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16151</th>\n",
       "      <td>ERR1230393</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>803265</td>\n",
       "      <td>722940</td>\n",
       "      <td>0.900002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16152</th>\n",
       "      <td>SRR2061198</td>\n",
       "      <td>Listeria monocytogenes</td>\n",
       "      <td>352405</td>\n",
       "      <td>317165</td>\n",
       "      <td>0.900001</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>16153 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               acc                    name  analyzed_spot_count  self_count  \\\n",
       "0       ERR1817002  Listeria monocytogenes               608593      594332   \n",
       "1       ERR1816986  Listeria monocytogenes               692180      673720   \n",
       "2       ERR1816981  Listeria monocytogenes               630954      613122   \n",
       "3       ERR1816996  Listeria monocytogenes               593342      575904   \n",
       "4       ERR1431120  Listeria monocytogenes              3249346     3153217   \n",
       "...            ...                     ...                  ...         ...   \n",
       "16148  SRR10665812  Listeria monocytogenes               510003      459009   \n",
       "16149   SRR2071980  Listeria monocytogenes              1336913     1203238   \n",
       "16150   SRR6038661  Listeria monocytogenes               749339      674410   \n",
       "16151   ERR1230393  Listeria monocytogenes               803265      722940   \n",
       "16152   SRR2061198  Listeria monocytogenes               352405      317165   \n",
       "\n",
       "       proportion  \n",
       "0        0.976567  \n",
       "1        0.973331  \n",
       "2        0.971738  \n",
       "3        0.970611  \n",
       "4        0.970416  \n",
       "...           ...  \n",
       "16148    0.900012  \n",
       "16149    0.900012  \n",
       "16150    0.900007  \n",
       "16151    0.900002  \n",
       "16152    0.900001  \n",
       "\n",
       "[16153 rows x 5 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT a.acc, a.name, info.analyzed_spot_count, a.self_count, a.self_count/info.analyzed_spot_count as proportion\n",
    "FROM `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis` a \n",
    "    JOIN `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis_info` info \n",
    "    ON a.acc = info.acc\n",
    "WHERE a.tax_id = 1639\n",
    "    AND a.self_count/info.analyzed_spot_count > .9\n",
    "    AND info.analyzed_spot_count > 100000\n",
    "ORDER BY proportion DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "6748fd71",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Query complete after 0.00s: 100%|██████████| 5/5 [00:00<00:00, 3950.18query/s]                        \n",
      "Downloading: 100%|██████████| 119/119 [00:00<00:00, 191.00rows/s]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>geo_loc_name_country_calc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>366105</td>\n",
       "      <td>USA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>196962</td>\n",
       "      <td>United Kingdom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>17804</td>\n",
       "      <td>Ireland</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15778</td>\n",
       "      <td>Spain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15653</td>\n",
       "      <td>Australia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>1</td>\n",
       "      <td>American Samoa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>1</td>\n",
       "      <td>Rwanda</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>1</td>\n",
       "      <td>Chad</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>1</td>\n",
       "      <td>Kyrgyzstan</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>1</td>\n",
       "      <td>Madagascar</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>119 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      count geo_loc_name_country_calc\n",
       "0    366105                       USA\n",
       "1    196962            United Kingdom\n",
       "2     17804                   Ireland\n",
       "3     15778                     Spain\n",
       "4     15653                 Australia\n",
       "..      ...                       ...\n",
       "114       1            American Samoa\n",
       "115       1                    Rwanda\n",
       "116       1                      Chad\n",
       "117       1                Kyrgyzstan\n",
       "118       1                Madagascar\n",
       "\n",
       "[119 rows x 2 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%bigquery\n",
    "SELECT count(meta.acc) as count, meta.geo_loc_name_country_calc\n",
    "FROM `nih-sra-datastore.sra.metadata` meta , `nih-sra-datastore.sra_tax_analysis_tool.tax_analysis` tax\n",
    "WHERE meta.acc = tax.acc\n",
    "    and tax.name = 'Coronaviridae'\n",
    "    and meta.geo_loc_name_country_calc is NOT NULL\n",
    "GROUP BY meta.geo_loc_name_country_calc\n",
    "ORDER by 1 DESC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71376c84",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
